// mediaforge/dist/cjs/process/spawn.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FFmpegSpawnError = void 0;
exports.spawnFFmpeg = spawnFFmpeg;
exports.runFFmpeg = runFFmpeg;
const child_process_1 = require("child_process");
const readline_1 = require("readline");
const events_js_1 = require("./events.js");
const progress_js_1 = require("./progress.js");
class FFmpegSpawnError extends Error {
    exitCode;
    signal;
    stderrOutput;
    constructor(exitCode, signal, stderrOutput) {
        super(`FFmpeg exited with code ${exitCode ?? signal ?? 'unknown'}:\n${stderrOutput.slice(-2000)}`);
        this.exitCode = exitCode;
        this.signal = signal;
        this.stderrOutput = stderrOutput;
        this.name = 'FFmpegSpawnError';
    }
}
exports.FFmpegSpawnError = FFmpegSpawnError;
/**
 * Spawn an ffmpeg child process and wire up all event handling.
 * Returns an FFmpegProcess immediately (process is already running).
 *
 * @example
 * const proc = spawnFFmpeg({ binary: 'ffmpeg', args: [...], parseProgress: true });
 * proc.emitter.on('progress', (info) => console.log(info.percent));
 * await new Promise((res, rej) => {
 *   proc.emitter.on('end', res);
 *   proc.emitter.on('error', rej);
 * });
 */
function spawnFFmpeg(opts) {
    const { binary, args, parseProgress = false, totalDurationUs, cwd } = opts;
    const child = (0, child_process_1.spawn)(binary, args, {
        stdio: ['pipe', 'pipe', 'pipe'],
        cwd,
    });
    const emitter = new events_js_1.FFmpegEmitter();
    const stderrLines = [];
    // Progress parser (only active if requested)
    const progressParser = parseProgress
        ? new progress_js_1.ProgressParser((info) => emitter.emit('progress', info), totalDurationUs)
        : null;
    emitter.emit('start', args);
    // Wire stderr → line-by-line
    if (child.stderr !== null) {
        const rl = (0, readline_1.createInterface)({ input: child.stderr, crlfDelay: Infinity });
        rl.on('line', (line) => {
            stderrLines.push(line);
            emitter.emit('stderr', line);
            progressParser?.push(line);
        });
    }
    // Handle process exit
    child.on('close', (code, signal) => {
        if (code === 0 || (code === null && signal === null)) {
            emitter.emit('end');
        }
        else {
            const stderr = stderrLines.join('\n');
            emitter.emit('error', new FFmpegSpawnError(code, signal, stderr));
        }
    });
    child.on('error', (err) => {
        emitter.emit('error', err);
    });
    return {
        emitter,
        child,
        stdin: child.stdin,
        stdout: child.stdout,
        kill(signal = 'SIGTERM') {
            child.kill(signal);
        },
    };
}
/**
 * Spawn ffmpeg and return a Promise that resolves when the process exits
 * successfully, or rejects with FFmpegSpawnError on failure.
 */
function runFFmpeg(opts) {
    return new Promise((resolve, reject) => {
        const proc = spawnFFmpeg(opts);
        proc.emitter.on('end', resolve);
        proc.emitter.on('error', reject);
    });
}
