// mediaforge/dist/esm/filters/audio/index.js
import { FilterChain, serializeNode } from '../../types/filters.js';
export function volume(chainOrOpts, opts) {
    const isStandalone = !(chainOrOpts instanceof FilterChain);
    const o = isStandalone ? chainOrOpts : opts;
    let node;
    if (typeof o === 'number' || typeof o === 'string') {
        node = { name: 'volume', positional: [String(o)], named: {} };
    }
    else {
        const named = {};
        if (o.replaygain !== undefined)
            named['replaygain'] = o.replaygain;
        if (o.eval !== undefined)
            named['eval'] = o.eval;
        if (o.precision !== undefined)
            named['precision'] = o.precision;
        node = { name: 'volume', positional: [String(o.volume)], named };
    }
    if (isStandalone)
        return serializeNode(node);
    return chainOrOpts.add(node);
}
export function loudnorm(chainOrOpts, opts) {
    const isStandalone = !(chainOrOpts instanceof FilterChain);
    const o = (isStandalone ? chainOrOpts : opts) ?? {};
    const named = {};
    if (o.i !== undefined)
        named['i'] = o.i;
    if (o.lra !== undefined)
        named['lra'] = o.lra;
    if (o.tp !== undefined)
        named['tp'] = o.tp;
    if (o.measured_i !== undefined)
        named['measured_i'] = o.measured_i;
    if (o.measured_tp !== undefined)
        named['measured_tp'] = o.measured_tp;
    if (o.measured_lra !== undefined)
        named['measured_lra'] = o.measured_lra;
    if (o.measured_thresh !== undefined)
        named['measured_thresh'] = o.measured_thresh;
    if (o.offset !== undefined)
        named['offset'] = o.offset;
    if (o.linear !== undefined)
        named['linear'] = o.linear ? 'true' : 'false';
    if (o.print_format !== undefined)
        named['print_format'] = o.print_format;
    if (o.dual_mono !== undefined)
        named['dual_mono'] = o.dual_mono ? 'true' : 'false';
    const node = { name: 'loudnorm', positional: [], named };
    if (isStandalone)
        return serializeNode(node);
    return chainOrOpts.add(node);
}
export function equalizer(chainOrOpts, opts) {
    const isStandalone = !(chainOrOpts instanceof FilterChain);
    const o = isStandalone ? chainOrOpts : opts;
    const named = {
        f: o.frequency,
        width: o.width,
        g: o.gain,
    };
    if (o.width_type !== undefined)
        named['width_type'] = o.width_type;
    if (o.channels !== undefined)
        named['c'] = o.channels;
    if (o.poles !== undefined)
        named['p'] = o.poles;
    if (o.mix !== undefined)
        named['mix'] = o.mix;
    const node = { name: 'equalizer', positional: [], named };
    if (isStandalone)
        return serializeNode(node);
    return chainOrOpts.add(node);
}
export function bass(chain, opts) {
    const named = { g: opts.gain };
    if (opts.frequency !== undefined)
        named['f'] = opts.frequency;
    if (opts.width !== undefined)
        named['w'] = opts.width;
    if (opts.width_type !== undefined)
        named['t'] = opts.width_type;
    return chain.add({ name: 'bass', positional: [], named });
}
export function treble(chain, opts) {
    const named = { g: opts.gain };
    if (opts.frequency !== undefined)
        named['f'] = opts.frequency;
    if (opts.width !== undefined)
        named['w'] = opts.width;
    if (opts.width_type !== undefined)
        named['t'] = opts.width_type;
    return chain.add({ name: 'treble', positional: [], named });
}
export function afade(chain, opts) {
    const named = { type: opts.type };
    if (opts.start_sample !== undefined)
        named['start_sample'] = opts.start_sample;
    if (opts.nb_samples !== undefined)
        named['nb_samples'] = opts.nb_samples;
    if (opts.start_time !== undefined)
        named['start_time'] = opts.start_time;
    if (opts.duration !== undefined)
        named['duration'] = opts.duration;
    if (opts.curve !== undefined)
        named['curve'] = opts.curve;
    return chain.add({ name: 'afade', positional: [], named });
}
// ─── Asetpts ──────────────────────────────────────────────────────────────────
/** Modify audio timestamps. Common: 'PTS-STARTPTS', 'NB_CONSUMED_SAMPLES*SR' */
export function asetpts(chain, expr) {
    return chain.add({ name: 'asetpts', positional: [expr], named: {} });
}
export function atrim(chain, opts) {
    const named = {};
    if (opts.start !== undefined)
        named['start'] = opts.start;
    if (opts.end !== undefined)
        named['end'] = opts.end;
    if (opts.duration !== undefined)
        named['duration'] = opts.duration;
    if (opts.start_pts !== undefined)
        named['start_pts'] = opts.start_pts;
    if (opts.end_pts !== undefined)
        named['end_pts'] = opts.end_pts;
    return chain.add({ name: 'atrim', positional: [], named });
}
// ─── Amerge ──────────────────────────────────────────────────────────────────
/** Merge N audio streams into a multi-channel stream */
export function amerge(chain, inputs = 2) {
    return chain.add({ name: 'amerge', positional: [], named: { inputs } });
}
export function amix(chain, opts = {}) {
    const named = {};
    if (opts.inputs !== undefined)
        named['inputs'] = opts.inputs;
    if (opts.duration !== undefined)
        named['duration'] = opts.duration;
    if (opts.dropout_transition !== undefined)
        named['dropout_transition'] = opts.dropout_transition;
    if (opts.normalize !== undefined)
        named['normalize'] = opts.normalize ? 1 : 0;
    if (opts.weights !== undefined)
        named['weights'] = opts.weights;
    return chain.add({ name: 'amix', positional: [], named });
}
// ─── Pan ─────────────────────────────────────────────────────────────────────
/**
 * Pan / remix audio channels using a channel layout expression.
 * @example pan(chain, 'stereo|c0=FL|c1=FR')
 * @example pan(chain, 'stereo|c0=0.5*c0+0.5*c1|c1=0.5*c0+0.5*c1')
 */
export function pan(chain, layoutExpr) {
    return chain.add({ name: 'pan', positional: [layoutExpr], named: {} });
}
// ─── Channelmap / Channelsplit ────────────────────────────────────────────────
export function channelmap(chain, map, channelLayout) {
    const named = { map };
    if (channelLayout !== undefined)
        named['channel_layout'] = channelLayout;
    return chain.add({ name: 'channelmap', positional: [], named });
}
export function channelsplit(chain, channelLayout) {
    const named = {};
    if (channelLayout !== undefined)
        named['channel_layout'] = channelLayout;
    return chain.add({ name: 'channelsplit', positional: [], named });
}
export function aresample(chain, opts) {
    if (typeof opts === 'number') {
        return chain.add({ name: 'aresample', positional: [opts], named: {} });
    }
    const named = {};
    if (opts.sampleRate !== undefined)
        named['sample_rate'] = opts.sampleRate;
    if (opts.resampler !== undefined)
        named['resampler'] = opts.resampler;
    if (opts.precision !== undefined)
        named['precision'] = opts.precision;
    return chain.add({ name: 'aresample', positional: [], named });
}
export function dynaudnorm(chain, opts = {}) {
    const named = {};
    if (opts.framelen !== undefined)
        named['f'] = opts.framelen;
    if (opts.gausssize !== undefined)
        named['g'] = opts.gausssize;
    if (opts.peak !== undefined)
        named['p'] = opts.peak;
    if (opts.maxgain !== undefined)
        named['m'] = opts.maxgain;
    if (opts.rms !== undefined)
        named['r'] = opts.rms;
    if (opts.compress !== undefined)
        named['s'] = opts.compress;
    if (opts.threshold !== undefined)
        named['t'] = opts.threshold;
    if (opts.coupling !== undefined)
        named['n'] = opts.coupling ? 1 : 0;
    if (opts.correctdc !== undefined)
        named['c'] = opts.correctdc ? 1 : 0;
    return chain.add({ name: 'dynaudnorm', positional: [], named });
}
export function compand(chain, opts = {}) {
    const named = {};
    if (opts.attacks !== undefined)
        named['attacks'] = opts.attacks;
    if (opts.decays !== undefined)
        named['decays'] = opts.decays;
    if (opts.points !== undefined)
        named['points'] = opts.points;
    if (opts.soft_knee !== undefined)
        named['soft_knee'] = opts.soft_knee;
    if (opts.gain !== undefined)
        named['gain'] = opts.gain;
    if (opts.delay !== undefined)
        named['delay'] = opts.delay;
    return chain.add({ name: 'compand', positional: [], named });
}
export function aecho(chain, opts = {}) {
    const named = {};
    if (opts.in_gain !== undefined)
        named['in_gain'] = opts.in_gain;
    if (opts.out_gain !== undefined)
        named['out_gain'] = opts.out_gain;
    if (opts.delays !== undefined)
        named['delays'] = opts.delays;
    if (opts.decays !== undefined)
        named['decays'] = opts.decays;
    return chain.add({ name: 'aecho', positional: [], named });
}
export function highpass(chain, opts) {
    if (typeof opts === 'number') {
        return chain.add({ name: 'highpass', positional: [], named: { f: opts } });
    }
    const named = { f: opts.frequency };
    if (opts.width !== undefined)
        named['w'] = opts.width;
    if (opts.width_type !== undefined)
        named['t'] = opts.width_type;
    if (opts.poles !== undefined)
        named['p'] = opts.poles;
    if (opts.mix !== undefined)
        named['m'] = opts.mix;
    return chain.add({ name: 'highpass', positional: [], named });
}
export function lowpass(chain, opts) {
    if (typeof opts === 'number') {
        return chain.add({ name: 'lowpass', positional: [], named: { f: opts } });
    }
    const named = { f: opts.frequency };
    if (opts.width !== undefined)
        named['w'] = opts.width;
    if (opts.width_type !== undefined)
        named['t'] = opts.width_type;
    if (opts.poles !== undefined)
        named['p'] = opts.poles;
    if (opts.mix !== undefined)
        named['m'] = opts.mix;
    return chain.add({ name: 'lowpass', positional: [], named });
}
// ─── Asplit ───────────────────────────────────────────────────────────────────
/** Split audio into N identical copies */
export function asplit(chain, n = 2) {
    return chain.add({ name: 'asplit', positional: [n], named: {} });
}
export function silencedetect(chain, opts = {}) {
    const named = {};
    if (opts.noise !== undefined)
        named['noise'] = opts.noise;
    if (opts.duration !== undefined)
        named['duration'] = opts.duration;
    if (opts.mono !== undefined)
        named['mono'] = opts.mono ? 1 : 0;
    return chain.add({ name: 'silencedetect', positional: [], named });
}
export function rubberband(chain, opts = {}) {
    const named = {};
    if (opts.tempo !== undefined)
        named['tempo'] = opts.tempo;
    if (opts.pitch !== undefined)
        named['pitch'] = opts.pitch;
    if (opts.transients !== undefined)
        named['transients'] = opts.transients;
    if (opts.detector !== undefined)
        named['detector'] = opts.detector;
    if (opts.phase !== undefined)
        named['phase'] = opts.phase;
    if (opts.window !== undefined)
        named['window'] = opts.window;
    if (opts.smoothing !== undefined)
        named['smoothing'] = opts.smoothing;
    if (opts.formant !== undefined)
        named['formant'] = opts.formant;
    if (opts.pitchq !== undefined)
        named['pitchq'] = opts.pitchq;
    if (opts.engine !== undefined)
        named['engine'] = opts.engine;
    return chain.add({ name: 'rubberband', positional: [], named });
}
export function atempo(chainOrOpts, factor) {
    const isStandalone = !(chainOrOpts instanceof FilterChain);
    let f;
    if (isStandalone) {
        f = typeof chainOrOpts === 'number' ? chainOrOpts : chainOrOpts.tempo;
        return serializeNode({ name: 'atempo', positional: [f], named: {} });
    }
    f = factor ?? 1;
    return chainOrOpts.add({ name: 'atempo', positional: [f], named: {} });
}
export function agate(chain, opts = {}) {
    const named = {};
    if (opts.threshold !== undefined)
        named['threshold'] = opts.threshold;
    if (opts.range !== undefined)
        named['range'] = opts.range;
    if (opts.attack !== undefined)
        named['attack'] = opts.attack;
    if (opts.release !== undefined)
        named['release'] = opts.release;
    if (opts.knee !== undefined)
        named['knee'] = opts.knee;
    if (opts.detection !== undefined)
        named['detection'] = opts.detection;
    if (opts.link !== undefined)
        named['link'] = opts.link;
    return chain.add({ name: 'agate', positional: [], named });
}
