// mediaforge/dist/cjs/FFmpeg.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GuardError = exports.FFmpegBuilder = exports.VersionError = void 0;
exports.ffmpeg = ffmpeg;
const binary_js_1 = require("./utils/binary.js");
const version_js_1 = require("./utils/version.js");
const args_js_1 = require("./utils/args.js");
const spawn_js_1 = require("./process/spawn.js");
const registry_js_1 = require("./codecs/registry.js");
const guards_js_1 = require("./compat/guards.js");
Object.defineProperty(exports, "GuardError", { enumerable: true, get: function () { return guards_js_1.GuardError; } });
class VersionError extends Error {
    constructor(feature, required, actual) {
        super(`"${feature}" requires FFmpeg v${required}+, but found v${actual}. ` +
            'Upgrade your FFmpeg installation.');
        this.name = 'VersionError';
    }
}
exports.VersionError = VersionError;
/**
 * Fluent FFmpeg builder.
 * All methods return `this` for chaining.
 * Call `.run()` to execute or `.spawn()` for event-based control.
 *
 * @example
 * await new FFmpegBuilder('input.mp4')
 *   .videoCodec('libx264')
 *   .videoBitrate('2M')
 *   .audioCodec('aac')
 *   .output('output.mp4')
 *   .run();
 */
class FFmpegBuilder {
    _binary;
    _version = null;
    _registry = null;
    _globalOpts = { overwrite: true };
    _inputs = [];
    _currentOutput = null;
    _outputs = [];
    constructor(input) {
        this._binary = (0, binary_js_1.resolveBinary)();
        if (input !== undefined) {
            const inputs = Array.isArray(input) ? input : [input];
            for (const i of inputs)
                this.input(i);
        }
    }
    // ─── Binary & Version ────────────────────────────────────────────────────
    /** Override the ffmpeg binary path (default: FFMPEG_PATH env or 'ffmpeg') */
    setBinary(path) {
        this._binary = path;
        this._version = null;
        this._registry = null; // reset capability cache too
        return this;
    }
    /** Get (and cache) the version info for the current binary. */
    async getVersion() {
        if (this._version === null) {
            this._version = (0, version_js_1.probeVersion)(this._binary);
        }
        return this._version;
    }
    /** Get (and cache) the capability registry for the current binary. */
    getRegistry() {
        if (this._registry === null) {
            this._registry = new registry_js_1.CapabilityRegistry(this._binary);
        }
        return this._registry;
    }
    // ─── Runtime guards ──────────────────────────────────────────────────────
    /**
     * Check if a codec is available in the current binary.
     * Does not throw — returns a GuardResult.
     */
    checkCodec(codec, direction = 'encode') {
        return (0, guards_js_1.guardCodec)(this.getRegistry(), codec, direction);
    }
    /**
     * Check if a hardware accelerator is available.
     */
    checkHwaccel(accelName) {
        return (0, guards_js_1.guardHwaccel)(this.getRegistry(), accelName);
    }
    /**
     * Check if a feature is available by version gate + capability.
     */
    async checkFeature(featureKey) {
        const v = await this.getVersion();
        return (0, guards_js_1.guardFeatureVersion)(v, featureKey);
    }
    /**
     * Select the best available video codec from a priority list.
     * Returns null if none are available.
     *
     * @example
     * const codec = await builder.selectVideoCodec([
     *   { codec: 'h264_nvenc', featureKey: 'nvenc' },
     *   { codec: 'h264_vaapi' },
     *   { codec: 'libx264' },
     * ]);
     * builder.videoCodec(codec ?? 'libx264');
     */
    async selectVideoCodec(candidates) {
        const v = await this.getVersion();
        return (0, guards_js_1.selectBestCodec)(v, this.getRegistry(), candidates);
    }
    /**
     * Select the best available hardware accelerator from a priority list.
     */
    selectHwaccel(candidates) {
        return (0, guards_js_1.selectBestHwaccel)(this.getRegistry(), candidates);
    }
    requireVersion(minMajor, feature) {
        if (this._version === null)
            return; // version not yet known — defer
        if (!(0, version_js_1.satisfiesVersion)(this._version, minMajor)) {
            throw new VersionError(feature, minMajor, this._version.major);
        }
    }
    // ─── Global Options ───────────────────────────────────────────────────────
    /** Overwrite output files without asking (default: true) */
    overwrite(yes = true) {
        this._globalOpts.overwrite = yes;
        if (!yes) {
            this._globalOpts.noOverwrite = true;
        }
        else {
            delete this._globalOpts.noOverwrite;
        }
        return this;
    }
    /** Set log level */
    logLevel(level) {
        this._globalOpts.logLevel = level;
        return this;
    }
    /** Enable progress reporting on stderr */
    enableProgress() {
        this._globalOpts.progress = true;
        return this;
    }
    // ─── Inputs ───────────────────────────────────────────────────────────────
    /** Add an input. Chainable input options follow. */
    input(path, opts = {}) {
        this._inputs.push({ path, opts });
        return this;
    }
    /** Seek within the last-added input (fast seek, before -i) */
    seekInput(position) {
        const last = this._inputs[this._inputs.length - 1];
        if (last !== undefined)
            last.opts.seekInput = position;
        return this;
    }
    /** Limit duration of the last-added input */
    inputDuration(duration) {
        const last = this._inputs[this._inputs.length - 1];
        if (last !== undefined)
            last.opts.duration = duration;
        return this;
    }
    /** Force format of the last-added input */
    inputFormat(format) {
        const last = this._inputs[this._inputs.length - 1];
        if (last !== undefined)
            last.opts.format = format;
        return this;
    }
    // ─── Output ───────────────────────────────────────────────────────────────
    ensureOutput() {
        if (this._currentOutput === null) {
            throw new Error('No output defined. Call .output("path") before codec/filter options.');
        }
        return this._currentOutput;
    }
    /** Add (or switch to) an output path. Call before codec/filter options. */
    output(path, opts = {}) {
        const entry = {
            path,
            opts,
            videoArgs: [],
            audioArgs: [],
            subtitleArgs: [],
            filterArgs: [],
        };
        this._outputs.push(entry);
        this._currentOutput = entry;
        return this;
    }
    // ─── Video Options ────────────────────────────────────────────────────────
    /** Set video codec, e.g. 'libx264', 'libx265', 'copy' */
    videoCodec(codec) {
        this.ensureOutput().videoArgs.push('-c:v', codec);
        return this;
    }
    /** Set video bitrate, e.g. '2M', '4000k' */
    videoBitrate(bitrate) {
        this.ensureOutput().videoArgs.push('-b:v', (0, args_js_1.toBitrate)(bitrate));
        return this;
    }
    /** Set video frame rate */
    fps(rate) {
        this.ensureOutput().videoArgs.push('-r', String(rate));
        return this;
    }
    /** Set output video size, e.g. '1280x720' or '1280:720' */
    size(widthXheight) {
        this.ensureOutput().videoArgs.push('-s', widthXheight);
        return this;
    }
    /** Set video filter chain (-vf). Accepts a filter string or a FilterChain / serialized filter function result. */
    videoFilter(filter) {
        this.ensureOutput().filterArgs.push('-vf', String(filter));
        return this;
    }
    /** Set pixel format */
    pixelFormat(fmt) {
        this.ensureOutput().videoArgs.push('-pix_fmt', fmt);
        return this;
    }
    /** Set CRF value (codec-dependent) */
    crf(value) {
        this.ensureOutput().videoArgs.push('-crf', String(value));
        return this;
    }
    /** Disable video stream */
    noVideo() {
        this.ensureOutput().videoArgs.push('-vn');
        return this;
    }
    // ─── Audio Options ────────────────────────────────────────────────────────
    /** Set audio codec, e.g. 'aac', 'libopus', 'copy' */
    audioCodec(codec) {
        this.ensureOutput().audioArgs.push('-c:a', codec);
        return this;
    }
    /** Set audio bitrate, e.g. '128k', '192k' */
    audioBitrate(bitrate) {
        this.ensureOutput().audioArgs.push('-b:a', (0, args_js_1.toBitrate)(bitrate));
        return this;
    }
    /** Set audio sample rate, e.g. 44100, 48000 */
    audioSampleRate(rate) {
        this.ensureOutput().audioArgs.push('-ar', String(rate));
        return this;
    }
    /** Set number of audio channels */
    audioChannels(channels) {
        this.ensureOutput().audioArgs.push('-ac', String(channels));
        return this;
    }
    /** Set audio filter chain (-af). Accepts a filter string or a FilterChain / serialized filter function result. */
    audioFilter(filter) {
        this.ensureOutput().filterArgs.push('-af', String(filter));
        return this;
    }
    /** Disable audio stream */
    noAudio() {
        this.ensureOutput().audioArgs.push('-an');
        return this;
    }
    // ─── Subtitle Options ─────────────────────────────────────────────────────
    /** Set subtitle codec */
    subtitleCodec(codec) {
        this.ensureOutput().subtitleArgs.push('-c:s', codec);
        return this;
    }
    /** Disable subtitle stream */
    noSubtitle() {
        this.ensureOutput().subtitleArgs.push('-sn');
        return this;
    }
    // ─── Output Format & Mapping ──────────────────────────────────────────────
    /** Force output format */
    outputFormat(format) {
        this.ensureOutput().opts.format = format;
        return this;
    }
    /** Add stream mapping, e.g. '0:v:0', '0:a:1' */
    map(specifier) {
        const out = this.ensureOutput();
        out.opts.map ??= [];
        out.opts.map.push(specifier);
        return this;
    }
    /** Limit output duration */
    duration(d) {
        this.ensureOutput().opts.duration = d;
        return this;
    }
    /** Seek in output (re-encode seek, accurate) */
    seekOutput(position) {
        this.ensureOutput().opts.seekOutput = position;
        return this;
    }
    /** Pass arbitrary extra args for the current output */
    addOutputOption(...args) {
        const out = this.ensureOutput();
        out.opts.extraArgs ??= [];
        out.opts.extraArgs.push(...args);
        return this;
    }
    /** Pass arbitrary extra args at the global level */
    addGlobalOption(...args) {
        this._globalOpts.extraArgs ??= [];
        this._globalOpts.extraArgs.push(...args);
        return this;
    }
    // ─── Hardware Acceleration ────────────────────────────────────────────────
    /** Enable hardware acceleration (-hwaccel). Probes version if cached. */
    hwAccel(accel, opts) {
        this._globalOpts.extraArgs ??= [];
        this._globalOpts.extraArgs.push('-hwaccel', accel);
        if (opts?.device !== undefined) {
            this._globalOpts.extraArgs.push('-hwaccel_device', opts.device);
        }
        return this;
    }
    // ─── Complex Filter ───────────────────────────────────────────────────────
    /** Set -filter_complex string directly (typed DSL coming in Phase 3) */
    complexFilter(filter) {
        this._globalOpts.extraArgs ??= [];
        this._globalOpts.extraArgs.push('-filter_complex', filter);
        return this;
    }
    // ─── Arg Assembly ─────────────────────────────────────────────────────────
    /** Build and return the full argument array. */
    buildArgs() {
        const args = [];
        // Global options
        args.push(...(0, args_js_1.buildGlobalArgs)(this._globalOpts));
        // Inputs
        for (const { path, opts } of this._inputs) {
            args.push(...(0, args_js_1.buildInputArgs)(opts));
            args.push('-i', path);
        }
        // Outputs
        for (const { path, opts, videoArgs, audioArgs, subtitleArgs, filterArgs } of this._outputs) {
            args.push(...videoArgs);
            args.push(...audioArgs);
            args.push(...subtitleArgs);
            args.push(...filterArgs);
            args.push(...(0, args_js_1.buildOutputArgs)(opts));
            args.push(path);
        }
        return args;
    }
    // ─── Execution ────────────────────────────────────────────────────────────
    /**
     * Spawn the process with full event-emitter control.
     * Useful for streaming progress or piping stdout.
     */
    spawn(opts) {
        const spawnOpts = {
            binary: this._binary,
            args: this.buildArgs(),
            parseProgress: opts?.parseProgress ?? this._globalOpts.progress === true,
        };
        if (opts?.totalDurationUs !== undefined)
            spawnOpts.totalDurationUs = opts.totalDurationUs;
        return (0, spawn_js_1.spawnFFmpeg)(spawnOpts);
    }
    /**
     * Run ffmpeg and return a Promise that resolves on success.
     * Rejects with FFmpegSpawnError on non-zero exit.
     */
    async run(opts) {
        const spawnOpts = {
            binary: this._binary,
            args: this.buildArgs(),
            parseProgress: opts?.parseProgress ?? this._globalOpts.progress === true,
        };
        if (opts?.totalDurationUs !== undefined)
            spawnOpts.totalDurationUs = opts.totalDurationUs;
        await (0, spawn_js_1.runFFmpeg)(spawnOpts);
    }
    /**
     * Probe the binary version and return a human-readable string.
     * Also caches the version for subsequent requireVersion() checks.
     */
    async versionString() {
        const v = await this.getVersion();
        return (0, version_js_1.formatVersion)(v);
    }
}
exports.FFmpegBuilder = FFmpegBuilder;
/**
 * Factory function — the primary public API entry point.
 *
 * @example
 * import { ffmpeg } from 'mediaforge';
 * await ffmpeg('input.mp4').videoCodec('libx264').output('out.mp4').run();
 */
function ffmpeg(input) {
    return new FFmpegBuilder(input);
}
