// mediaforge/dist/cjs/codecs/audio.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.aacToArgs = aacToArgs;
exports.opusToArgs = opusToArgs;
exports.libOpusToArgs = opusToArgs;
exports.mp3ToArgs = mp3ToArgs;
exports.libMp3LameToArgs = mp3ToArgs;
exports.flacToArgs = flacToArgs;
exports.ac3ToArgs = ac3ToArgs;
exports.alacToArgs = alacToArgs;
exports.eac3ToArgs = eac3ToArgs;
exports.truehdToArgs = truehdToArgs;
exports.vorbisToArgs = vorbisToArgs;
exports.wavpackToArgs = wavpackToArgs;
exports.pcmToArgs = pcmToArgs;
exports.mp2ToArgs = mp2ToArgs;
/**
 * Serialize AacOptions to ffmpeg argument array.
 */
function aacToArgs(opts) {
    const args = ['-c:a', 'aac'];
    if (opts.aacCoder !== undefined)
        args.push('-aac_coder', opts.aacCoder);
    if (opts.vbr !== undefined)
        args.push('-vbr', String(opts.vbr));
    if (opts.bitrate !== undefined)
        args.push('-b:a', `${opts.bitrate}k`);
    if (opts.sampleRate !== undefined)
        args.push('-ar', String(opts.sampleRate));
    if (opts.channels !== undefined)
        args.push('-ac', String(opts.channels));
    if (opts.channelLayout !== undefined)
        args.push('-channel_layout', opts.channelLayout);
    return args;
}
/**
 * Serialize LibOpusOptions to ffmpeg argument array.
 */
function opusToArgs(opts) {
    const args = ['-c:a', 'libopus'];
    if (opts.bitrate !== undefined)
        args.push('-b:a', `${opts.bitrate}k`);
    if (opts.vbr !== undefined)
        args.push('-vbr', opts.vbr);
    if (opts.application !== undefined)
        args.push('-application', opts.application);
    if (opts.frameDuration !== undefined)
        args.push('-frame_duration', String(opts.frameDuration));
    if (opts.compressionLevel !== undefined)
        args.push('-compression_level', String(opts.compressionLevel));
    if (opts.packetLoss !== undefined)
        args.push('-packet_loss', String(opts.packetLoss));
    if (opts.fec !== undefined)
        args.push('-fec', opts.fec ? '1' : '0');
    if (opts.dtx !== undefined)
        args.push('-dtx', opts.dtx ? '1' : '0');
    if (opts.channels !== undefined)
        args.push('-ac', String(opts.channels));
    if (opts.sampleRate !== undefined)
        args.push('-ar', String(opts.sampleRate));
    return args;
}
/**
 * Serialize LibMp3LameOptions to ffmpeg argument array.
 */
function mp3ToArgs(opts) {
    const args = ['-c:a', 'libmp3lame'];
    if (opts.qscale !== undefined)
        args.push('-q:a', String(opts.qscale));
    if (opts.bitrate !== undefined)
        args.push('-b:a', `${opts.bitrate}k`);
    if (opts.compressionLevel !== undefined)
        args.push('-compression_level', String(opts.compressionLevel));
    if (opts.jointStereo !== undefined)
        args.push('-joint_stereo', opts.jointStereo ? '1' : '0');
    if (opts.abr === true)
        args.push('-abr', '1');
    if (opts.sampleRate !== undefined)
        args.push('-ar', String(opts.sampleRate));
    if (opts.channels !== undefined)
        args.push('-ac', String(opts.channels));
    return args;
}
/**
 * Serialize FlacOptions to ffmpeg argument array.
 */
function flacToArgs(opts) {
    const args = ['-c:a', 'flac'];
    if (opts.compressionLevel !== undefined)
        args.push('-compression_level', String(opts.compressionLevel));
    if (opts.lpcOrder !== undefined)
        args.push('-lpc_order', String(opts.lpcOrder));
    if (opts.lpcCoeffPrecision !== undefined)
        args.push('-lpc_coeff_precision', String(opts.lpcCoeffPrecision));
    if (opts.predictionOrderMethod !== undefined)
        args.push('-prediction_order_method', opts.predictionOrderMethod);
    if (opts.minPartitionOrder !== undefined)
        args.push('-min_partition_order', String(opts.minPartitionOrder));
    if (opts.maxPartitionOrder !== undefined)
        args.push('-max_partition_order', String(opts.maxPartitionOrder));
    return args;
}
// ─── ac3ToArgs ────────────────────────────────────────────────────────────────
function ac3ToArgs(opts) {
    const args = ['-c:a', 'ac3'];
    if (opts.bitrate !== undefined)
        args.push('-b:a', `${opts.bitrate}k`);
    if (opts.dialogueLevel !== undefined)
        args.push('-dialnorm', String(opts.dialogueLevel));
    if (opts.centerMixLevel !== undefined)
        args.push('-center_mixlev', String(opts.centerMixLevel));
    if (opts.surroundMixLevel !== undefined)
        args.push('-surround_mixlev', String(opts.surroundMixLevel));
    if (opts.audioCodingMode !== undefined)
        args.push('-acmod', String(opts.audioCodingMode));
    return args;
}
function alacToArgs(opts = {}) {
    const args = ['-c:a', 'alac'];
    if (opts.minPredictionOrder !== undefined)
        args.push('-min_prediction_order', String(opts.minPredictionOrder));
    if (opts.maxPredictionOrder !== undefined)
        args.push('-max_prediction_order', String(opts.maxPredictionOrder));
    return args;
}
function eac3ToArgs(opts = {}) {
    const args = ['-c:a', 'eac3'];
    if (opts.bitrate !== undefined)
        args.push('-b:a', `${opts.bitrate}k`);
    if (opts.dialNorm !== undefined)
        args.push('-dialnorm', String(opts.dialNorm));
    if (opts.centerMixLevel !== undefined)
        args.push('-center_mixlev', String(opts.centerMixLevel));
    if (opts.surroundMixLevel !== undefined)
        args.push('-surround_mixlev', String(opts.surroundMixLevel));
    return args;
}
function truehdToArgs(opts = {}) {
    const args = ['-c:a', 'truehd'];
    if (opts.sampleRate !== undefined)
        args.push('-ar', String(opts.sampleRate));
    if (opts.channelLayout !== undefined)
        args.push('-channel_layout', opts.channelLayout);
    return args;
}
function vorbisToArgs(opts = {}) {
    const args = ['-c:a', 'libvorbis'];
    if (opts.qscale !== undefined)
        args.push('-q:a', String(opts.qscale));
    else if (opts.bitrate !== undefined)
        args.push('-b:a', `${opts.bitrate}k`);
    if (opts.minrate !== undefined)
        args.push('-minrate', `${opts.minrate}k`);
    if (opts.maxrate !== undefined)
        args.push('-maxrate', `${opts.maxrate}k`);
    return args;
}
function wavpackToArgs(opts = {}) {
    const args = ['-c:a', 'wavpack'];
    if (opts.quality === 'lossless' || opts.quality === undefined) {
        // default is lossless
    }
    else if (typeof opts.quality === 'number') {
        args.push('-compression_level', String(opts.quality));
    }
    if (opts.bitrate !== undefined)
        args.push('-b:a', `${opts.bitrate}k`);
    if (opts.extra !== undefined)
        args.push('-extra', String(opts.extra));
    return args;
}
function pcmToArgs(format, opts = {}) {
    const args = ['-c:a', format];
    if (opts.sampleRate !== undefined)
        args.push('-ar', String(opts.sampleRate));
    if (opts.channels !== undefined)
        args.push('-ac', String(opts.channels));
    return args;
}
function mp2ToArgs(opts = {}) {
    const args = ['-c:a', 'mp2'];
    if (opts.bitrate !== undefined)
        args.push('-b:a', `${opts.bitrate}k`);
    if (opts.sampleRate !== undefined)
        args.push('-ar', String(opts.sampleRate));
    return args;
}
