// mediaforge/dist/cjs/codecs/hardware.js
"use strict";
/**
 * Hardware acceleration codec option interfaces.
 * Covers everything in the v7 (NVENC/VAAPI) and v8 (MediaCodec/Vulkan) dumps.
 * Runtime availability is always probed — never hardcoded by platform.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.nvencToArgs = nvencToArgs;
exports.vaapiToArgs = vaapiToArgs;
exports.mediacodecToArgs = mediacodecToArgs;
exports.vulkanToArgs = vulkanToArgs;
exports.qsvToArgs = qsvToArgs;
exports.mediacodecVideoToArgs = mediacodecVideoToArgs;
exports.vulkanVideoToArgs = vulkanVideoToArgs;
/** Build ffmpeg arg list for NVENC */
function nvencToArgs(opts, codec = 'h264_nvenc') {
    const args = ['-c:v', codec];
    if (opts.preset !== undefined)
        args.push('-preset', opts.preset);
    if (opts.tune !== undefined)
        args.push('-tune', opts.tune);
    if (opts.rcMode !== undefined)
        args.push('-rc', opts.rcMode);
    if (opts.constqp !== undefined)
        args.push('-constqp', String(opts.constqp));
    if (opts.bitrate !== undefined)
        args.push('-b:v', `${opts.bitrate}k`);
    if (opts.maxrate !== undefined)
        args.push('-maxrate', `${opts.maxrate}k`);
    if (opts.bufsize !== undefined)
        args.push('-bufsize', `${opts.bufsize}k`);
    if (opts.cq !== undefined)
        args.push('-cq', String(opts.cq));
    if (opts.qpI !== undefined)
        args.push('-qp:v', String(opts.qpI));
    if (opts.bFrames !== undefined)
        args.push('-bf', String(opts.bFrames));
    if (opts.refs !== undefined)
        args.push('-refs', String(opts.refs));
    if (opts.gopSize !== undefined)
        args.push('-g', String(opts.gopSize));
    if (opts.aqStrength !== undefined)
        args.push('-aq-strength', String(opts.aqStrength));
    if (opts.temporalAq !== undefined)
        args.push('-temporal-aq', opts.temporalAq ? '1' : '0');
    if (opts.weightedPred !== undefined)
        args.push('-weighted_pred', opts.weightedPred ? '1' : '0');
    if (opts.level !== undefined)
        args.push('-level', opts.level);
    if (opts.profile !== undefined)
        args.push('-profile:v', opts.profile);
    if (opts.rcLookahead !== undefined)
        args.push('-rc-lookahead', String(opts.rcLookahead));
    if (opts.gpuDevice !== undefined)
        args.push('-gpu', String(opts.gpuDevice));
    return args;
}
/** Build ffmpeg arg list for VAAPI */
function vaapiToArgs(opts, codec = 'h264_vaapi') {
    const args = ['-c:v', codec];
    if (opts.qp !== undefined)
        args.push('-qp', String(opts.qp));
    if (opts.bitrate !== undefined)
        args.push('-b:v', `${opts.bitrate}k`);
    if (opts.maxrate !== undefined)
        args.push('-maxrate', `${opts.maxrate}k`);
    if (opts.keyintMax !== undefined)
        args.push('-g', String(opts.keyintMax));
    if (opts.bFrames !== undefined)
        args.push('-bf', String(opts.bFrames));
    if (opts.refs !== undefined)
        args.push('-refs', String(opts.refs));
    if (opts.profile !== undefined)
        args.push('-profile:v', opts.profile);
    if (opts.level !== undefined)
        args.push('-level', String(opts.level));
    if (opts.rcMode !== undefined)
        args.push('-rc_mode', opts.rcMode);
    return args;
}
/** Build ffmpeg arg list for MediaCodec */
function mediacodecToArgs(opts, codec = 'h264_mediacodec') {
    const args = ['-c:v', codec];
    if (opts.ndkCodec === true)
        args.push('-ndk_codec', '1');
    if (opts.bitrate !== undefined)
        args.push('-b:v', `${opts.bitrate}k`);
    if (opts.iFrameInterval !== undefined)
        args.push('-i_qfactor', String(opts.iFrameInterval));
    if (opts.level !== undefined)
        args.push('-level', String(opts.level));
    return args;
}
/** Build ffmpeg arg list for Vulkan */
function vulkanToArgs(opts, codec = 'h264_vulkan') {
    const args = ['-c:v', codec];
    if (opts.bitrate !== undefined)
        args.push('-b:v', `${opts.bitrate}k`);
    if (opts.qp !== undefined)
        args.push('-qp', String(opts.qp));
    if (opts.gopSize !== undefined)
        args.push('-g', String(opts.gopSize));
    if (opts.bFrames !== undefined)
        args.push('-bf', String(opts.bFrames));
    return args;
}
/** Build ffmpeg arg list for QSV */
function qsvToArgs(opts, codec = 'h264_qsv') {
    const args = ['-c:v', codec];
    if (opts.preset !== undefined)
        args.push('-preset', opts.preset);
    if (opts.globalQuality !== undefined)
        args.push('-global_quality', String(opts.globalQuality));
    if (opts.bitrate !== undefined)
        args.push('-b:v', `${opts.bitrate}k`);
    if (opts.maxrate !== undefined)
        args.push('-maxrate', `${opts.maxrate}k`);
    if (opts.gopSize !== undefined)
        args.push('-g', String(opts.gopSize));
    if (opts.bFrames !== undefined)
        args.push('-bf', String(opts.bFrames));
    if (opts.lowLatency === true)
        args.push('-low_delay_brc', '1');
    return args;
}
function mediacodecVideoToArgs(opts, codec = 'h264_mediacodec') {
    const args = ['-c:v', codec];
    if (opts.bitrate !== undefined)
        args.push('-b:v', `${opts.bitrate}k`);
    if (opts.quality !== undefined)
        args.push('-q:v', String(opts.quality));
    if (opts.operatingRate !== undefined)
        args.push('-operating_rate', String(opts.operatingRate));
    if (opts.profile !== undefined)
        args.push('-profile:v', String(opts.profile));
    if (opts.level !== undefined)
        args.push('-level', String(opts.level));
    return args;
}
function vulkanVideoToArgs(opts, codec = 'h264_vulkan') {
    const args = ['-c:v', codec];
    if (opts.crf !== undefined)
        args.push('-crf', String(opts.crf));
    if (opts.bitrate !== undefined)
        args.push('-b:v', `${opts.bitrate}k`);
    return args;
}
