// mediaforge/dist/cjs/codecs/registry.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CapabilityRegistry = void 0;
exports.getDefaultRegistry = getDefaultRegistry;
const child_process_1 = require("child_process");
/**
 * Runtime capability registry.
 * Probes the installed ffmpeg binary once and caches the results.
 * Zero hardcoding — the binary self-describes what it supports.
 */
class CapabilityRegistry {
    binary;
    _codecs = null;
    _filters = null;
    _formats = null;
    _hwaccels = null;
    _encoders = null;
    constructor(binary) {
        this.binary = binary;
    }
    // ─── Codecs ───────────────────────────────────────────────────────────────
    get codecs() {
        if (this._codecs === null) {
            this._codecs = this._probeCodecs();
        }
        return this._codecs;
    }
    hasCodec(name) {
        // Check codec family names (e.g. 'h264') AND individual encoder/decoder names (e.g. 'libx264')
        return this.codecs.has(name) || this.encoders.has(name);
    }
    get encoders() {
        // Encoder names are populated when codecs are probed (from "(encoders: ...)" parentheticals).
        // Trigger codec probe if not yet done — this also fills _encoders.
        if (this._encoders === null) {
            void this.codecs; // triggers _probeCodecs which sets this._encoders
        }
        return this._encoders ?? new Set();
    }
    canEncode(codec) {
        // Check codec-level encode flag (e.g. 'h264') AND individual encoder name (e.g. 'libx264')
        return (this.codecs.get(codec)?.flags.encode ?? false) || this.encoders.has(codec);
    }
    canDecode(codec) {
        return this.codecs.get(codec)?.flags.decode ?? false;
    }
    // ─── Filters ──────────────────────────────────────────────────────────────
    get filters() {
        if (this._filters === null) {
            this._filters = this._probeFilters();
        }
        return this._filters;
    }
    hasFilter(name) {
        return this.filters.has(name);
    }
    // ─── Formats ──────────────────────────────────────────────────────────────
    get formats() {
        if (this._formats === null) {
            this._formats = this._probeFormats();
        }
        return this._formats;
    }
    hasFormat(name) {
        return this.formats.has(name);
    }
    // ─── Hardware Accels ──────────────────────────────────────────────────────
    get hwaccels() {
        if (this._hwaccels === null) {
            this._hwaccels = this._probeHwaccels();
        }
        return this._hwaccels;
    }
    hasHwaccel(name) {
        return this.hwaccels.has(name);
    }
    // ─── Private probe methods ────────────────────────────────────────────────
    _exec(args) {
        return (0, child_process_1.execFileSync)(this.binary, args, {
            stdio: ['ignore', 'pipe', 'pipe'],
            encoding: 'utf8',
        });
    }
    /**
     * Parse `ffmpeg -codecs` output.
     * Each codec line: " DEV.LS name  description (encoders: libx264 h264_nvenc ...) (decoders: ...)"
     * Also populates the _encoders set from the "(encoders: ...)" parenthetical,
     * mapping individual encoder names (libx264, h264_nvenc) to the codec family (h264).
     */
    _probeCodecs() {
        const map = new Map();
        // Reset encoder names so they are rebuilt alongside the codec map
        this._encoders = new Set();
        let output;
        try {
            output = this._exec(['-codecs', '-hide_banner']);
        }
        catch {
            return map;
        }
        const lines = output.split('\n');
        let pastHeader = false;
        for (const line of lines) {
            if (!pastHeader) {
                if (line.trimStart().startsWith('-----'))
                    pastHeader = true;
                continue;
            }
            // Format: " DEV.LS codec_name   description"
            const match = /^ ([D.])([E.])([VASDT])([I.])([L.])([S.])\s+(\S+)\s+(.*)$/.exec(line);
            if (match === null)
                continue;
            const [, d, e, t, i, l, s, name, description] = match;
            if (name === undefined || description === undefined)
                continue;
            const typeChar = t ?? '.';
            const typeMap = {
                V: 'video', A: 'audio', S: 'subtitle', D: 'data', T: 'attachment',
            };
            map.set(name, {
                name,
                description: description.trim(),
                flags: {
                    decode: d === 'D',
                    encode: e === 'E',
                    type: typeMap[typeChar] ?? 'data',
                    intraOnly: i === 'I',
                    lossy: l === 'L',
                    lossless: s === 'S',
                },
            });
            // Extract individual encoder names from "(encoders: libx264 h264_nvenc ...)"
            // These are distinct from the codec family name and are what users pass to -c:v
            const encMatch = /\(encoders?:([^)]+)\)/.exec(description);
            if (encMatch?.[1]) {
                for (const enc of encMatch[1].trim().split(/\s+/)) {
                    if (enc)
                        this._encoders.add(enc);
                }
            }
        }
        return map;
    }
    /**
     * Parse `ffmpeg -filters` output.
     * Filter line format: " T.S name   description"
     */
    _probeFilters() {
        const map = new Map();
        let output;
        try {
            output = this._exec(['-filters', '-hide_banner']);
        }
        catch {
            return map;
        }
        for (const line of output.split('\n')) {
            // v7: "T.S name  desc", v8: "TS name  desc" (2-char flags)
            const match = /^ [T.][S.][C.]?\s+(\S+)\s+(.+)$/.exec(line);
            if (match === null)
                continue;
            const [, name, description] = match;
            if (name === undefined || description === undefined)
                continue;
            map.set(name, {
                name,
                description: description.trim(),
                timeline: false,
                sliceBased: false,
            });
        }
        return map;
    }
    /**
     * Parse `ffmpeg -formats` output.
     * Format line format: " DE name   description"
     */
    _probeFormats() {
        const map = new Map();
        let output;
        try {
            output = this._exec(['-formats', '-hide_banner']);
        }
        catch {
            return map;
        }
        let pastHeader = false;
        for (const line of output.split('\n')) {
            if (!pastHeader) {
                if (line.includes('--'))
                    pastHeader = true;
                continue;
            }
            const match = /^ ([D.])([E.])\s+(\S+)\s+(.+)$/.exec(line);
            if (match === null)
                continue;
            const [, d, e, name, description] = match;
            if (name === undefined || description === undefined)
                continue;
            // Format name can be comma-separated aliases: "matroska,webm"
            for (const alias of name.split(',')) {
                map.set(alias.trim(), {
                    name: alias.trim(),
                    description: description.trim(),
                    demux: d === 'D',
                    mux: e === 'E',
                });
            }
        }
        return map;
    }
    /**
     * Parse `ffmpeg -hwaccels` output.
     */
    _probeHwaccels() {
        const set = new Set();
        let output;
        try {
            output = this._exec(['-hwaccels', '-hide_banner']);
        }
        catch {
            return set;
        }
        let started = false;
        for (const line of output.split('\n')) {
            const trimmed = line.trim();
            if (trimmed === 'Hardware acceleration methods:') {
                started = true;
                continue;
            }
            if (started && trimmed.length > 0)
                set.add(trimmed);
        }
        return set;
    }
    /** Invalidate all cached capability data (e.g. after setBinary) */
    invalidate() {
        this._codecs = null;
        this._filters = null;
        this._formats = null;
        this._hwaccels = null;
    }
}
exports.CapabilityRegistry = CapabilityRegistry;
/** Global default registry (lazy, uses FFMPEG_PATH / system ffmpeg) */
let _defaultRegistry = null;
function getDefaultRegistry(binary = 'ffmpeg') {
    if (_defaultRegistry === null) {
        _defaultRegistry = new CapabilityRegistry(binary);
    }
    return _defaultRegistry;
}
