// mediaforge/dist/esm/helpers/normalize.js
import { spawnFFmpeg, runFFmpeg } from '../process/spawn.js';
import { resolveBinary } from '../utils/binary.js';
/**
 * Normalize audio loudness (EBU R128 / ITU-R BS.1770).
 *
 * @example
 * const result = await normalizeAudio({ input: 'raw.mp4', output: 'norm.mp4', targetI: -16 });
 */
export async function normalizeAudio(opts) {
    const { input, output, targetI = -23, targetLra = 7, targetTp = -2, twoPass = true, videoCodec = 'copy', binary = resolveBinary(), } = opts;
    if (!twoPass) {
        await runFFmpeg({
            binary,
            args: ['-y', '-i', input, '-c:v', videoCodec, '-af', `loudnorm=i=${targetI}:lra=${targetLra}:tp=${targetTp}`, output],
        });
        return { inputI: targetI, inputLra: targetLra, inputTp: targetTp, inputThresh: targetI - 10, targetOffset: 0 };
    }
    // Pass 1: measure
    const stderrLines = [];
    await new Promise((resolve, reject) => {
        const proc = spawnFFmpeg({
            binary,
            args: ['-y', '-i', input, '-af', `loudnorm=i=${targetI}:lra=${targetLra}:tp=${targetTp}:print_format=json`, '-f', 'null', '-'],
        });
        proc.emitter.on('stderr', (line) => stderrLines.push(line));
        proc.emitter.on('end', resolve);
        proc.emitter.on('error', reject);
    });
    let measured = { inputI: targetI, inputLra: targetLra, inputTp: targetTp, inputThresh: targetI - 10, targetOffset: 0 };
    const jsonStr = stderrLines.join('\n');
    const match = jsonStr.match(/\{[\s\S]*?"input_i"[\s\S]*?\}/);
    if (match) {
        try {
            const j = JSON.parse(match[0]);
            measured = {
                inputI: parseFloat(j.input_i),
                inputLra: parseFloat(j.input_lra),
                inputTp: parseFloat(j.input_tp),
                inputThresh: parseFloat(j.input_thresh),
                targetOffset: parseFloat(j.target_offset),
            };
        }
        catch { /* use defaults */ }
    }
    // Pass 2: apply
    const applyFilter = `loudnorm=i=${targetI}:lra=${targetLra}:tp=${targetTp}:measured_i=${measured.inputI}:measured_lra=${measured.inputLra}:measured_tp=${measured.inputTp}:measured_thresh=${measured.inputThresh}:offset=${measured.targetOffset}:linear=true`;
    await runFFmpeg({
        binary,
        args: ['-y', '-i', input, '-c:v', videoCodec, '-af', applyFilter, output],
    });
    return measured;
}
/**
 * Adjust volume by a multiplier or dB value.
 * @example
 * await adjustVolume({ input: 'in.mp4', output: 'out.mp4', volume: '0.5' });
 * await adjustVolume({ input: 'in.mp4', output: 'out.mp4', volume: '6dB' });
 */
export async function adjustVolume(opts) {
    const { input, output, volume, videoCodec = 'copy', binary = resolveBinary() } = opts;
    await runFFmpeg({
        binary,
        args: ['-y', '-i', input, '-c:v', videoCodec, '-af', `volume=${volume}`, output],
    });
}
// ─── Arg builders ─────────────────────────────────────────────────────────────
export function buildLoudnormFilter(targetI, targetLra, targetTp, measured) {
    let filter = `loudnorm=i=${targetI}:lra=${targetLra}:tp=${targetTp}`;
    if (measured) {
        filter += `:measured_i=${measured.inputI}:measured_lra=${measured.inputLra}` +
            `:measured_tp=${measured.inputTp}:measured_thresh=${measured.inputThresh}` +
            `:offset=${measured.targetOffset}:linear=true`;
    }
    return filter;
}
