// mediaforge/dist/esm/helpers/screenshots.js
import * as fs from 'fs';
import * as path from 'path';
import { spawnFFmpeg } from '../process/spawn.js';
import { resolveBinary } from '../utils/binary.js';
function toSeconds(ts) {
    if (typeof ts === 'number')
        return ts;
    const parts = ts.split(':').map(Number);
    if (parts.length === 3)
        return parts[0] * 3600 + parts[1] * 60 + parts[2];
    if (parts.length === 2)
        return parts[0] * 60 + parts[1];
    return Number(ts);
}
/**
 * Extract screenshots from a video file.
 *
 * @example
 * // 5 evenly-spaced screenshots
 * const { files } = await screenshots({ input: 'video.mp4', folder: './thumbs', count: 5 });
 *
 * @example
 * // At specific timestamps
 * const { files } = await screenshots({
 *   input: 'video.mp4', folder: './thumbs',
 *   timestamps: ['00:00:05', '00:00:30', 90],
 * });
 */
export async function screenshots(opts) {
    const { input, folder, count, timestamps, filename = 'screenshot_%04d.png', size, binary = resolveBinary(), } = opts;
    if (!fs.existsSync(folder))
        fs.mkdirSync(folder, { recursive: true });
    if (timestamps && timestamps.length > 0) {
        // Per-timestamp mode: one ffmpeg call per timestamp
        const files = [];
        for (let i = 0; i < timestamps.length; i++) {
            const ts = timestamps[i];
            const ext = path.extname(filename) || '.png';
            const base = path.basename(filename, ext);
            // Replace %04d style format with index
            const outName = base.replace(/%0?\d*d/, String(i + 1).padStart(4, '0')) + ext;
            const outPath = path.join(folder, outName);
            const args = ['-y', '-ss', String(toSeconds(ts)), '-i', input, '-vframes', '1'];
            if (size)
                args.push('-s', size);
            args.push(outPath);
            await new Promise((resolve, reject) => {
                const proc = spawnFFmpeg({ binary, args });
                proc.emitter.on('end', resolve);
                proc.emitter.on('error', reject);
            });
            files.push(outPath);
        }
        return { files };
    }
    if (count && count > 0) {
        // Count mode: use fps filter to extract N evenly-spaced frames
        // Get duration first via ffprobe
        const { probe } = await import('../probe/ffprobe.js');
        const info = probe(input);
        const duration = info.format?.duration ? parseFloat(info.format.duration) : null;
        if (duration === null || duration <= 0) {
            throw new Error(`Cannot determine duration of "${input}" for count-based screenshots`);
        }
        const interval = duration / (count + 1);
        const times = Array.from({ length: count }, (_, i) => interval * (i + 1));
        const { count: _count, ...restOpts } = opts;
        return screenshots({ ...restOpts, timestamps: times });
    }
    // Fallback: single screenshot at 0
    const { count: _count2, ...restOpts2 } = opts;
    return screenshots({ ...restOpts2, timestamps: [0] });
}
export async function frameToBuffer(opts) {
    const { input, timestamp = 0, format = 'png', size, binary = resolveBinary(), } = opts;
    const args = [
        '-y',
        '-ss', String(toSeconds(timestamp)),
        '-i', input,
        '-vframes', '1',
        '-f', 'image2pipe',
        '-vcodec', format === 'mjpeg' ? 'mjpeg' : format,
    ];
    if (size)
        args.push('-s', size);
    args.push('pipe:1');
    return new Promise((resolve, reject) => {
        const proc = spawnFFmpeg({ binary, args });
        const chunks = [];
        if (proc.stdout) {
            proc.stdout.on('data', (chunk) => chunks.push(chunk));
        }
        proc.emitter.on('end', () => resolve(Buffer.concat(chunks)));
        proc.emitter.on('error', reject);
    });
}
// ─── Arg builders ─────────────────────────────────────────────────────────────
export function buildScreenshotArgs(input, outputPath, timestamp, size) {
    const args = ['-y', '-ss', String(timestamp), '-i', input, '-vframes', '1'];
    if (size)
        args.push('-s', size);
    args.push(outputPath);
    return args;
}
export function buildFrameBufferArgs(input, timestamp, format, size) {
    const args = ['-y', '-ss', String(timestamp), '-i', input, '-vframes', '1',
        '-f', 'image2pipe', '-vcodec', format === 'mjpeg' ? 'mjpeg' : format];
    if (size)
        args.push('-s', size);
    args.push('pipe:1');
    return args;
}
export function buildTimestampFilename(pattern, index, ext) {
    const base = path.basename(pattern, ext);
    return base.replace(/%0?\d*d/, String(index + 1).padStart(4, '0')) + ext;
}
