// mediaforge/dist/cjs/types/filters.js
"use strict";
/**
 * Core type system for the filter DSL.
 * Covers -vf, -af, and -filter_complex graph building.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.FilterChain = void 0;
exports.serializeNode = serializeNode;
exports.pad = pad;
exports.serializeLink = serializeLink;
/**
 * Serialize a FilterNode to its ffmpeg string representation.
 * e.g. { name:'scale', positional:[1280,720], named:{flags:'lanczos'} }
 *   → "scale=1280:720:flags=lanczos"
 */
function serializeNode(node) {
    const parts = [];
    // Positional args join with ':'
    if (node.positional.length > 0) {
        parts.push(node.positional.map(String).join(':'));
    }
    // Named args join as key=value pairs, then ':' between them
    const named = Object.entries(node.named);
    if (named.length > 0) {
        parts.push(...named.map(([k, v]) => `${k}=${String(v)}`));
    }
    if (parts.length === 0)
        return node.name;
    return `${node.name}=${parts.join(':')}`;
}
/**
 * A complete filter chain (for -vf or -af).
 * Filters are applied in order, output of each feeds the next.
 */
class FilterChain {
    nodes = [];
    /** Add a raw FilterNode */
    add(node) {
        this.nodes.push(node);
        return this;
    }
    /** Add a raw filter string (escape hatch) */
    raw(filterStr) {
        // Stored as a special raw node
        this.nodes.push({ name: '__raw__', positional: [filterStr], named: {} });
        return this;
    }
    /** Serialize the full chain to a comma-separated string */
    toString() {
        return this.nodes
            .map((n) => (n.name === '__raw__' ? String(n.positional[0] ?? '') : serializeNode(n)))
            .join(',');
    }
    /** Number of filters in the chain */
    get length() {
        return this.nodes.length;
    }
    /** Get all nodes (for testing/inspection) */
    getNodes() {
        return this.nodes;
    }
}
exports.FilterChain = FilterChain;
/**
 * Create a GraphPad from a label string.
 */
function pad(label) {
    return {
        label,
        toString: () => `[${label}]`,
    };
}
/**
 * Serialize a GraphLink to its filter_complex segment string.
 * e.g. [0:v][1:v]overlay=x=10:y=10[out]
 */
function serializeLink(link) {
    const ins = link.inputs.map(String).join('');
    const filter = serializeNode(link.filter);
    const outs = link.outputs.map(String).join('');
    return `${ins}${filter}${outs}`;
}
