// mediaforge/dist/cjs/utils/binary.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BinaryNotExecutableError = exports.BinaryNotFoundError = void 0;
exports.resolveBinary = resolveBinary;
exports.resolveProbe = resolveProbe;
exports.validateBinary = validateBinary;
exports.isBinaryAvailable = isBinaryAvailable;
const child_process_1 = require("child_process");
const fs_1 = require("fs");
class BinaryNotFoundError extends Error {
    constructor(binary) {
        super(`FFmpeg binary not found: "${binary}". ` +
            'Set the path via ffmpeg().setBinary() or the FFMPEG_PATH environment variable.');
        this.name = 'BinaryNotFoundError';
    }
}
exports.BinaryNotFoundError = BinaryNotFoundError;
class BinaryNotExecutableError extends Error {
    constructor(binary) {
        super(`FFmpeg binary is not executable: "${binary}"`);
        this.name = 'BinaryNotExecutableError';
    }
}
exports.BinaryNotExecutableError = BinaryNotExecutableError;
/**
 * Resolve the ffmpeg binary path.
 * Priority: explicit path → FFMPEG_PATH env → 'ffmpeg' on PATH
 */
function resolveBinary(explicit) {
    const candidate = explicit ?? process.env['FFMPEG_PATH'] ?? 'ffmpeg';
    return candidate;
}
/**
 * Resolve the ffprobe binary path.
 * Priority: explicit path → FFPROBE_PATH env → 'ffprobe' on PATH
 */
function resolveProbe(explicit) {
    return explicit ?? process.env['FFPROBE_PATH'] ?? 'ffprobe';
}
/**
 * Validate that a binary exists and is executable.
 * Throws BinaryNotFoundError or BinaryNotExecutableError on failure.
 */
function validateBinary(binaryPath) {
    // For absolute/relative paths, check directly.
    // For plain names (no path separator), rely on exec to find them on PATH.
    if (binaryPath.includes('/') || binaryPath.includes('\\')) {
        try {
            (0, fs_1.accessSync)(binaryPath, fs_1.constants.F_OK);
        }
        catch {
            throw new BinaryNotFoundError(binaryPath);
        }
        try {
            (0, fs_1.accessSync)(binaryPath, fs_1.constants.X_OK);
        }
        catch {
            throw new BinaryNotExecutableError(binaryPath);
        }
    }
    else {
        // Plain name — do a quick version check to confirm it resolves
        try {
            (0, child_process_1.execFileSync)(binaryPath, ['-version'], { stdio: 'pipe' });
        }
        catch (err) {
            const code = err.code;
            if (code === 'ENOENT')
                throw new BinaryNotFoundError(binaryPath);
            throw new BinaryNotExecutableError(binaryPath);
        }
    }
}
/**
 * Return true if the binary can be found and executed.
 */
function isBinaryAvailable(binaryPath) {
    try {
        validateBinary(binaryPath);
        return true;
    }
    catch {
        return false;
    }
}
