// mediaforge/dist/cjs/filters/video/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.scale = scale;
exports.crop = crop;
exports.pad = pad;
exports.overlay = overlay;
exports.drawtext = drawtext;
exports.fps = fps;
exports.setpts = setpts;
exports.trim = trim;
exports.format = format;
exports.setsar = setsar;
exports.setdar = setdar;
exports.vflip = vflip;
exports.hflip = hflip;
exports.rotate = rotate;
exports.transpose = transpose;
exports.unsharp = unsharp;
exports.gblur = gblur;
exports.boxblur = boxblur;
exports.eq = eq;
exports.hue = hue;
exports.colorbalance = colorbalance;
exports.yadif = yadif;
exports.hqdn3d = hqdn3d;
exports.nlmeans = nlmeans;
exports.thumbnail = thumbnail;
exports.select = select;
exports.concat = concat;
exports.split = split;
exports.tile = tile;
exports.colorkey = colorkey;
exports.chromakey = chromakey;
exports.subtitles = subtitles;
exports.avgblurVulkan = avgblurVulkan;
exports.nlmeansVulkan = nlmeansVulkan;
exports.fade = fade;
exports.zoompan = zoompan;
const filters_js_1 = require("../../types/filters.js");
function scale(chainOrOpts, opts) {
    const isStandalone = !(chainOrOpts instanceof filters_js_1.FilterChain);
    const o = isStandalone ? chainOrOpts : opts;
    const w = o.w ?? o.width ?? '-1';
    const h = o.h ?? o.height ?? '-1';
    const named = {};
    if (o.flags !== undefined)
        named['flags'] = o.flags;
    if (o.force_original_aspect_ratio !== undefined)
        named['force_original_aspect_ratio'] = o.force_original_aspect_ratio;
    if (o.force_divisible_by !== undefined)
        named['force_divisible_by'] = o.force_divisible_by;
    if (o.eval !== undefined)
        named['eval'] = o.eval;
    const node = { name: 'scale', positional: [String(w), String(h)], named };
    if (isStandalone)
        return (0, filters_js_1.serializeNode)(node);
    return chainOrOpts.add(node);
}
function crop(chainOrOpts, opts) {
    const isStandalone = !(chainOrOpts instanceof filters_js_1.FilterChain);
    const o = isStandalone ? chainOrOpts : opts;
    const w = o.w ?? o.width ?? 'iw';
    const h = o.h ?? o.height ?? 'ih';
    const named = {};
    if (o.x !== undefined)
        named['x'] = o.x;
    if (o.y !== undefined)
        named['y'] = o.y;
    if (o.keep_aspect !== undefined)
        named['keep_aspect'] = o.keep_aspect ? 1 : 0;
    if (o.exact !== undefined)
        named['exact'] = o.exact ? 1 : 0;
    const node = { name: 'crop', positional: [String(w), String(h)], named };
    if (isStandalone)
        return (0, filters_js_1.serializeNode)(node);
    return chainOrOpts.add(node);
}
function pad(chain, opts) {
    const named = {};
    if (opts.x !== undefined)
        named['x'] = opts.x;
    if (opts.y !== undefined)
        named['y'] = opts.y;
    if (opts.color !== undefined)
        named['color'] = opts.color;
    if (opts.aspect !== undefined)
        named['aspect'] = opts.aspect;
    return chain.add({
        name: 'pad',
        positional: [String(opts.width), String(opts.height)],
        named,
    });
}
function overlay(chainOrOpts, opts) {
    const isStandalone = !(chainOrOpts instanceof filters_js_1.FilterChain);
    const o = isStandalone ? chainOrOpts : opts;
    const named = {};
    if (o.eval !== undefined)
        named['eval'] = o.eval;
    if (o.shortest !== undefined)
        named['shortest'] = o.shortest ? 1 : 0;
    if (o.format !== undefined)
        named['format'] = o.format;
    if (o.alpha !== undefined)
        named['alpha'] = o.alpha;
    const node = { name: 'overlay', positional: [String(o.x), String(o.y)], named };
    if (isStandalone)
        return (0, filters_js_1.serializeNode)(node);
    return chainOrOpts.add(node);
}
function drawtext(chainOrOpts, opts) {
    const isStandalone = !(chainOrOpts instanceof filters_js_1.FilterChain);
    const o = isStandalone ? chainOrOpts : opts;
    if (isStandalone) {
        // standalone - build and serialize immediately
        return _drawtextImpl(new filters_js_1.FilterChain(), o).toString();
    }
    return _drawtextImpl(chainOrOpts, o);
}
function _drawtextImpl(chain, opts) {
    const named = {};
    if (opts.text !== undefined)
        named['text'] = opts.text;
    if (opts.textfile !== undefined)
        named['textfile'] = opts.textfile;
    if (opts.fontfile !== undefined)
        named['fontfile'] = opts.fontfile;
    if (opts.font !== undefined)
        named['font'] = opts.font;
    if (opts.fontsize !== undefined)
        named['fontsize'] = opts.fontsize;
    if (opts.fontcolor !== undefined)
        named['fontcolor'] = opts.fontcolor;
    if (opts.fontcolor_expr !== undefined)
        named['fontcolor_expr'] = opts.fontcolor_expr;
    if (opts.boxcolor !== undefined)
        named['boxcolor'] = opts.boxcolor;
    if (opts.box !== undefined)
        named['box'] = opts.box ? 1 : 0;
    if (opts.boxborderw !== undefined)
        named['boxborderw'] = opts.boxborderw;
    if (opts.x !== undefined)
        named['x'] = opts.x;
    if (opts.y !== undefined)
        named['y'] = opts.y;
    if (opts.shadowcolor !== undefined)
        named['shadowcolor'] = opts.shadowcolor;
    if (opts.shadowx !== undefined)
        named['shadowx'] = opts.shadowx;
    if (opts.shadowy !== undefined)
        named['shadowy'] = opts.shadowy;
    if (opts.borderw !== undefined)
        named['borderw'] = opts.borderw;
    if (opts.bordercolor !== undefined)
        named['bordercolor'] = opts.bordercolor;
    if (opts.line_spacing !== undefined)
        named['line_spacing'] = opts.line_spacing;
    if (opts.enable !== undefined)
        named['enable'] = opts.enable;
    if (opts.expansion !== undefined)
        named['expansion'] = opts.expansion;
    if (opts.text_align !== undefined)
        named['text_align'] = opts.text_align;
    return chain.add({ name: 'drawtext', positional: [], named });
}
function fps(chain, opts) {
    const named = {};
    if (opts.round !== undefined)
        named['round'] = opts.round;
    if (opts.eof_action !== undefined)
        named['eof_action'] = opts.eof_action;
    return chain.add({ name: 'fps', positional: [String(opts.fps)], named });
}
// ─── Setpts ───────────────────────────────────────────────────────────────────
/** Modify video timestamps. Common: 'PTS-STARTPTS', '2*PTS', 'PTS/2' */
function setpts(chain, expr) {
    return chain.add({ name: 'setpts', positional: [expr], named: {} });
}
function trim(chain, opts) {
    const named = {};
    if (opts.start !== undefined)
        named['start'] = opts.start;
    if (opts.end !== undefined)
        named['end'] = opts.end;
    if (opts.duration !== undefined)
        named['duration'] = opts.duration;
    if (opts.start_frame !== undefined)
        named['start_frame'] = opts.start_frame;
    if (opts.end_frame !== undefined)
        named['end_frame'] = opts.end_frame;
    return chain.add({ name: 'trim', positional: [], named });
}
// ─── Format / Setsar / Setdar ─────────────────────────────────────────────────
/** Convert pixel format */
function format(chain, pixFmt) {
    const fmts = Array.isArray(pixFmt) ? pixFmt.join('|') : pixFmt;
    return chain.add({ name: 'format', positional: [fmts], named: {} });
}
/** Set sample aspect ratio */
function setsar(chain, ratio, max) {
    const named = {};
    if (max !== undefined)
        named['max'] = max;
    return chain.add({ name: 'setsar', positional: [ratio], named });
}
/** Set display aspect ratio */
function setdar(chain, ratio, max) {
    const named = {};
    if (max !== undefined)
        named['max'] = max;
    return chain.add({ name: 'setdar', positional: [ratio], named });
}
// ─── Flip / Rotate ────────────────────────────────────────────────────────────
function vflip(chain) {
    return chain.add({ name: 'vflip', positional: [], named: {} });
}
function hflip(chain) {
    return chain.add({ name: 'hflip', positional: [], named: {} });
}
function rotate(chain, opts) {
    const named = {
        a: String(opts.angle),
    };
    if (opts.ow !== undefined)
        named['ow'] = opts.ow;
    if (opts.oh !== undefined)
        named['oh'] = opts.oh;
    if (opts.fillcolor !== undefined)
        named['fillcolor'] = opts.fillcolor;
    if (opts.bilinear !== undefined)
        named['bilinear'] = opts.bilinear ? 1 : 0;
    return chain.add({ name: 'rotate', positional: [], named });
}
function transpose(chain, dir) {
    return chain.add({ name: 'transpose', positional: [dir], named: {} });
}
function unsharp(chain, opts = {}) {
    const named = {};
    if (opts.lx !== undefined)
        named['lx'] = opts.lx;
    if (opts.ly !== undefined)
        named['ly'] = opts.ly;
    if (opts.la !== undefined)
        named['la'] = opts.la;
    if (opts.cx !== undefined)
        named['cx'] = opts.cx;
    if (opts.cy !== undefined)
        named['cy'] = opts.cy;
    if (opts.ca !== undefined)
        named['ca'] = opts.ca;
    return chain.add({ name: 'unsharp', positional: [], named });
}
function gblur(chain, opts = {}) {
    const named = {};
    if (opts.sigma !== undefined)
        named['sigma'] = opts.sigma;
    if (opts.steps !== undefined)
        named['steps'] = opts.steps;
    if (opts.planes !== undefined)
        named['planes'] = opts.planes;
    if (opts.sigmaV !== undefined)
        named['sigmaV'] = opts.sigmaV;
    return chain.add({ name: 'gblur', positional: [], named });
}
function boxblur(chain, opts = {}) {
    const named = {};
    if (opts.luma_radius !== undefined)
        named['luma_radius'] = opts.luma_radius;
    if (opts.luma_power !== undefined)
        named['luma_power'] = opts.luma_power;
    if (opts.chroma_radius !== undefined)
        named['chroma_radius'] = opts.chroma_radius;
    if (opts.chroma_power !== undefined)
        named['chroma_power'] = opts.chroma_power;
    return chain.add({ name: 'boxblur', positional: [], named });
}
function eq(chain, opts) {
    const named = {};
    if (opts.brightness !== undefined)
        named['brightness'] = opts.brightness;
    if (opts.contrast !== undefined)
        named['contrast'] = opts.contrast;
    if (opts.saturation !== undefined)
        named['saturation'] = opts.saturation;
    if (opts.gamma !== undefined)
        named['gamma'] = opts.gamma;
    if (opts.gamma_weight !== undefined)
        named['gamma_weight'] = opts.gamma_weight;
    if (opts.gamma_r !== undefined)
        named['gamma_r'] = opts.gamma_r;
    if (opts.gamma_g !== undefined)
        named['gamma_g'] = opts.gamma_g;
    if (opts.gamma_b !== undefined)
        named['gamma_b'] = opts.gamma_b;
    if (opts.eval !== undefined)
        named['eval'] = opts.eval;
    return chain.add({ name: 'eq', positional: [], named });
}
function hue(chain, opts) {
    const named = {};
    if (opts.h !== undefined)
        named['h'] = opts.h;
    if (opts.s !== undefined)
        named['s'] = opts.s;
    if (opts.b !== undefined)
        named['b'] = opts.b;
    return chain.add({ name: 'hue', positional: [], named });
}
function colorbalance(chain, opts) {
    const named = {};
    for (const [k, v] of Object.entries(opts)) {
        if (v !== undefined)
            named[k] = v;
    }
    return chain.add({ name: 'colorbalance', positional: [], named });
}
function yadif(chain, opts = {}) {
    const named = {};
    if (opts.mode !== undefined)
        named['mode'] = opts.mode;
    if (opts.parity !== undefined)
        named['parity'] = opts.parity;
    if (opts.deint !== undefined)
        named['deint'] = opts.deint;
    return chain.add({ name: 'yadif', positional: [], named });
}
function hqdn3d(chain, opts = {}) {
    // hqdn3d positional: luma_spatial:chroma_spatial:luma_tmp:chroma_tmp
    const pos = [];
    if (opts.s0 !== undefined)
        pos.push(opts.s0);
    if (opts.s1 !== undefined)
        pos.push(opts.s1);
    if (opts.s2 !== undefined)
        pos.push(opts.s2);
    if (opts.s3 !== undefined)
        pos.push(opts.s3);
    return chain.add({ name: 'hqdn3d', positional: pos, named: {} });
}
function nlmeans(chain, opts = {}) {
    const named = {};
    if (opts.h !== undefined)
        named['h'] = opts.h;
    if (opts.r !== undefined)
        named['r'] = opts.r;
    if (opts.p !== undefined)
        named['p'] = opts.p;
    if (opts.s !== undefined)
        named['s'] = opts.s;
    return chain.add({ name: 'nlmeans', positional: [], named });
}
// ─── Thumbnail ────────────────────────────────────────────────────────────────
/** Select the best frame from a group of N frames for thumbnail generation */
function thumbnail(chain, n = 100) {
    return chain.add({ name: 'thumbnail', positional: [n], named: {} });
}
// ─── Select ──────────────────────────────────────────────────────────────────
/** Select specific frames using an expression, e.g. 'not(mod(n,10))' */
function select(chain, expr) {
    return chain.add({ name: 'select', positional: [expr], named: {} });
}
function concat(chain, opts = {}) {
    const named = {};
    if (opts.n !== undefined)
        named['n'] = opts.n;
    if (opts.v !== undefined)
        named['v'] = opts.v;
    if (opts.a !== undefined)
        named['a'] = opts.a;
    if (opts.unsafe === true)
        named['unsafe'] = 1;
    return chain.add({ name: 'concat', positional: [], named });
}
// ─── Split ────────────────────────────────────────────────────────────────────
/** Split a video into N identical copies for multi-output filter graphs */
function split(chain, n = 2) {
    return chain.add({ name: 'split', positional: [n], named: {} });
}
function tile(chain, opts) {
    const named = {};
    if (opts.nb_frames !== undefined)
        named['nb_frames'] = opts.nb_frames;
    if (opts.margin !== undefined)
        named['margin'] = opts.margin;
    if (opts.padding !== undefined)
        named['padding'] = opts.padding;
    if (opts.color !== undefined)
        named['color'] = opts.color;
    if (opts.overlap !== undefined)
        named['overlap'] = opts.overlap;
    return chain.add({ name: 'tile', positional: [opts.layout], named });
}
function colorkey(chain, opts) {
    const named = {};
    if (opts.similarity !== undefined)
        named['similarity'] = opts.similarity;
    if (opts.blend !== undefined)
        named['blend'] = opts.blend;
    return chain.add({ name: 'colorkey', positional: [opts.color], named });
}
function chromakey(chain, opts) {
    const named = {};
    if (opts.similarity !== undefined)
        named['similarity'] = opts.similarity;
    if (opts.blend !== undefined)
        named['blend'] = opts.blend;
    return chain.add({ name: 'chromakey', positional: [opts.color], named });
}
function subtitles(chain, opts) {
    const named = { filename: opts.filename };
    if (opts.si !== undefined)
        named['si'] = opts.si;
    if (opts.original_size !== undefined)
        named['original_size'] = opts.original_size;
    if (opts.charenc !== undefined)
        named['charenc'] = opts.charenc;
    if (opts.force_style !== undefined)
        named['force_style'] = opts.force_style;
    return chain.add({ name: 'subtitles', positional: [], named });
}
function avgblurVulkan(chain, opts = {}) {
    const named = {};
    if (opts.sizeX !== undefined)
        named['sizeX'] = opts.sizeX;
    if (opts.sizeY !== undefined)
        named['sizeY'] = opts.sizeY;
    if (opts.planes !== undefined)
        named['planes'] = opts.planes;
    return chain.add({ name: 'avgblur_vulkan', positional: [], named });
}
function nlmeansVulkan(chain, opts = {}) {
    const named = {};
    if (opts.h !== undefined)
        named['h'] = opts.h;
    if (opts.r !== undefined)
        named['r'] = opts.r;
    if (opts.p !== undefined)
        named['p'] = opts.p;
    return chain.add({ name: 'nlmeans_vulkan', positional: [], named });
}
function fade(chainOrOpts, opts) {
    const isStandalone = !(chainOrOpts instanceof filters_js_1.FilterChain);
    const o = isStandalone ? chainOrOpts : opts;
    const named = { type: o.type };
    if (o.start_frame !== undefined)
        named['start_frame'] = o.start_frame;
    if (o.nb_frames !== undefined)
        named['nb_frames'] = o.nb_frames;
    if (o.start_time !== undefined)
        named['start_time'] = o.start_time;
    if (o.duration !== undefined)
        named['duration'] = o.duration;
    if (o.color !== undefined)
        named['color'] = o.color;
    if (o.alpha !== undefined)
        named['alpha'] = o.alpha ? 1 : 0;
    const node = { name: 'fade', positional: [], named };
    if (isStandalone)
        return (0, filters_js_1.serializeNode)(node);
    return chainOrOpts.add(node);
}
function zoompan(chain, opts = {}) {
    const named = {};
    if (opts.zoom !== undefined)
        named['zoom'] = opts.zoom;
    if (opts.x !== undefined)
        named['x'] = opts.x;
    if (opts.y !== undefined)
        named['y'] = opts.y;
    if (opts.d !== undefined)
        named['d'] = opts.d;
    if (opts.s !== undefined)
        named['s'] = opts.s;
    if (opts.fps !== undefined)
        named['fps'] = opts.fps;
    return chain.add({ name: 'zoompan', positional: [], named });
}
