// mediaforge/dist/esm/filters/complex.js
import { FilterChain, serializeNode, } from '../types/filters.js';
let _labelCounter = 0;
function autoLabel(prefix = 'p') {
    return `${prefix}${_labelCounter++}`;
}
/** Reset the auto-label counter (useful in tests) */
export function resetLabelCounter() {
    _labelCounter = 0;
}
/**
 * A stream in the graph with a typed label.
 * Wraps a pad label and tracks its type for validation.
 */
export class GraphStream {
    label;
    type;
    constructor(label, type) {
        this.label = label;
        this.type = type;
    }
    toString() {
        return `[${this.label}]`;
    }
}
/**
 * A node being built in the filter graph.
 * Supports chaining multiple filters on the same stream.
 */
export class GraphNode {
    graph;
    inputs;
    chain = [];
    constructor(graph, inputs) {
        this.graph = graph;
        this.inputs = inputs;
    }
    /** Apply a raw filter string */
    filter(name, positional = [], named = {}) {
        this.chain.push({ name, positional, named });
        return this;
    }
    // ─── Video filters (convenience wrappers) ──────────────────────────────────
    scale(width, height, opts = {}) {
        this.chain.push({ name: 'scale', positional: [String(width), String(height)], named: opts });
        return this;
    }
    crop(width, height, x, y) {
        const named = {};
        if (x !== undefined)
            named['x'] = x;
        if (y !== undefined)
            named['y'] = y;
        this.chain.push({ name: 'crop', positional: [String(width), String(height)], named });
        return this;
    }
    pad(width, height, x, y, color) {
        const named = {};
        if (x !== undefined)
            named['x'] = x;
        if (y !== undefined)
            named['y'] = y;
        if (color !== undefined)
            named['color'] = color;
        this.chain.push({ name: 'pad', positional: [String(width), String(height)], named });
        return this;
    }
    fps(rate) {
        this.chain.push({ name: 'fps', positional: [String(rate)], named: {} });
        return this;
    }
    format(pixFmt) {
        this.chain.push({ name: 'format', positional: [pixFmt], named: {} });
        return this;
    }
    setpts(expr) {
        this.chain.push({ name: 'setpts', positional: [expr], named: {} });
        return this;
    }
    vflip() {
        this.chain.push({ name: 'vflip', positional: [], named: {} });
        return this;
    }
    hflip() {
        this.chain.push({ name: 'hflip', positional: [], named: {} });
        return this;
    }
    colorkey(color, similarity = 0.1, blend = 0.0) {
        this.chain.push({ name: 'colorkey', positional: [color], named: { similarity, blend } });
        return this;
    }
    // ─── Audio filters (convenience wrappers) ──────────────────────────────────
    volume(v) {
        this.chain.push({ name: 'volume', positional: [String(v)], named: {} });
        return this;
    }
    atempo(factor) {
        this.chain.push({ name: 'atempo', positional: [factor], named: {} });
        return this;
    }
    asetpts(expr) {
        this.chain.push({ name: 'asetpts', positional: [expr], named: {} });
        return this;
    }
    loudnorm(i = -23, lra = 7, tp = -2) {
        this.chain.push({ name: 'loudnorm', positional: [], named: { i, lra, tp } });
        return this;
    }
    // ─── Output labeling ───────────────────────────────────────────────────────
    /**
     * Assign output labels to this node and register the link in the graph.
     * Returns GraphStream[] that can be used as inputs to subsequent filters.
     */
    out(...labels) {
        if (labels.length === 0)
            labels = [autoLabel()];
        const outputs = labels.map((l) => new GraphStream(l, 'unknown'));
        this._commit(outputs);
        return outputs;
    }
    /** Shorthand: single output with auto-generated label */
    outAuto(type = 'unknown') {
        const label = autoLabel(type === 'video' ? 'v' : 'a');
        const stream = new GraphStream(label, type);
        this._commit([stream]);
        return stream;
    }
    /** Register this as a final output (no output label needed — used with -map) */
    mapOut(label) {
        const lbl = label ?? autoLabel('out');
        const stream = new GraphStream(lbl, 'unknown');
        this._commit([stream]);
        return stream;
    }
    _commit(outputs) {
        // Serialize the chain as a sequence of comma-joined filters
        const filterStr = this.chain.map(serializeNode).join(',');
        const inputLabels = this.inputs.map((s) => `[${s.label}]`).join('');
        const outputLabels = outputs.map((s) => `[${s.label}]`).join('');
        this.graph._addLink(`${inputLabels}${filterStr}${outputLabels}`);
    }
}
/**
 * Multi-input filter node (overlay, amerge, amix, concat, etc.)
 */
export class MultiInputNode {
    graph;
    streams;
    constructor(graph, streams) {
        this.graph = graph;
        this.streams = streams;
    }
    overlay(x = 0, y = 0, opts = {}) {
        const named = { ...opts };
        const node = new GraphNode(this.graph, this.streams);
        return node.filter('overlay', [String(x), String(y)], named);
    }
    amerge(inputs) {
        const named = {};
        if (inputs !== undefined)
            named['inputs'] = inputs;
        return new GraphNode(this.graph, this.streams).filter('amerge', [], named);
    }
    amix(inputs, duration) {
        const named = {};
        if (inputs !== undefined)
            named['inputs'] = inputs;
        if (duration !== undefined)
            named['duration'] = duration;
        return new GraphNode(this.graph, this.streams).filter('amix', [], named);
    }
    concat(n, v, a) {
        return new GraphNode(this.graph, this.streams).filter('concat', [], { n, v, a });
    }
    hstack(inputs, shortest) {
        const named = {};
        if (inputs !== undefined)
            named['inputs'] = inputs;
        if (shortest !== undefined)
            named['shortest'] = shortest ? 1 : 0;
        return new GraphNode(this.graph, this.streams).filter('hstack', [], named);
    }
    vstack(inputs, shortest) {
        const named = {};
        if (inputs !== undefined)
            named['inputs'] = inputs;
        if (shortest !== undefined)
            named['shortest'] = shortest ? 1 : 0;
        return new GraphNode(this.graph, this.streams).filter('vstack', [], named);
    }
}
/**
 * The main filter_complex graph builder.
 *
 * @example
 * const g = new FilterGraph();
 *
 * // Scale the main video
 * const [v] = g.from('[0:v]').scale(1280, 720).out('scaled');
 *
 * // Prepare audio: normalize + fade in
 * const [a] = g.from('[0:a]').loudnorm(23, 7, -2).out('audio');
 *
 * // Build final filter_complex string
 * console.log(g.toString());
 * // → "[0:v]scale=1280:720[scaled];[0:a]loudnorm=i=23:lra=7:tp=-2[audio]"
 */
export class FilterGraph {
    _links = [];
    /** @internal Called by GraphNode._commit() */
    _addLink(link) {
        this._links.push(link);
    }
    /**
     * Start building a filter chain from one or more input stream refs.
     * @example g.from('[0:v]')   // from input 0 video
     * @example g.from('[logo]')  // from a previously labeled stream
     */
    from(...refs) {
        const streams = refs.map((r) => typeof r === 'string'
            ? new GraphStream(r.replace(/^\[|\]$/g, ''), 'unknown')
            : r);
        return new GraphNode(this, streams);
    }
    /**
     * Start building a multi-input filter (overlay, amerge, concat…).
     */
    merge(...refs) {
        const streams = refs.map((r) => typeof r === 'string'
            ? new GraphStream(r.replace(/^\[|\]$/g, ''), 'unknown')
            : r);
        return new MultiInputNode(this, streams);
    }
    /**
     * Serialize the complete filter_complex string (semicolon-separated links).
     */
    toString() {
        return this._links.join(';');
    }
    /** How many links are in the graph */
    get size() {
        return this._links.length;
    }
}
// ─── Simple chain builders (for -vf / -af) ────────────────────────────────────
/**
 * Build a typed video filter chain for use with .videoFilter().
 *
 * @example
 * const vf = videoFilterChain()
 *   .scale(1280, 720)
 *   .unsharp({ la: 1.0 })
 *   .toString();
 * ffmpeg('in.mp4').output('out.mp4').videoFilter(vf).run();
 */
export class VideoFilterChain {
    chain = new FilterChain();
    raw(str) { this.chain.raw(str); return this; }
    scale(width, height, flags) {
        const named = {};
        if (flags !== undefined)
            named['flags'] = flags;
        this.chain.add({ name: 'scale', positional: [String(width), String(height)], named });
        return this;
    }
    crop(w, h, x, y) {
        const named = {};
        if (x !== undefined)
            named['x'] = x;
        if (y !== undefined)
            named['y'] = y;
        this.chain.add({ name: 'crop', positional: [String(w), String(h)], named });
        return this;
    }
    pad(w, h, x, y, color) {
        const named = {};
        if (x !== undefined)
            named['x'] = x;
        if (y !== undefined)
            named['y'] = y;
        if (color !== undefined)
            named['color'] = color;
        this.chain.add({ name: 'pad', positional: [String(w), String(h)], named });
        return this;
    }
    fps(rate) {
        this.chain.add({ name: 'fps', positional: [String(rate)], named: {} });
        return this;
    }
    setpts(expr) {
        this.chain.add({ name: 'setpts', positional: [expr], named: {} });
        return this;
    }
    vflip() { this.chain.add({ name: 'vflip', positional: [], named: {} }); return this; }
    hflip() { this.chain.add({ name: 'hflip', positional: [], named: {} }); return this; }
    format(fmt) {
        this.chain.add({ name: 'format', positional: [fmt], named: {} });
        return this;
    }
    unsharp(lx = 5, ly = 5, la = 1.0) {
        this.chain.add({ name: 'unsharp', positional: [], named: { lx, ly, la } });
        return this;
    }
    eq(brightness, contrast, saturation, gamma) {
        const named = {};
        if (brightness !== undefined)
            named['brightness'] = brightness;
        if (contrast !== undefined)
            named['contrast'] = contrast;
        if (saturation !== undefined)
            named['saturation'] = saturation;
        if (gamma !== undefined)
            named['gamma'] = gamma;
        this.chain.add({ name: 'eq', positional: [], named });
        return this;
    }
    drawtext(text, x, y, fontsize, color) {
        const named = { text, x, y };
        if (fontsize !== undefined)
            named['fontsize'] = fontsize;
        if (color !== undefined)
            named['fontcolor'] = color;
        this.chain.add({ name: 'drawtext', positional: [], named });
        return this;
    }
    yadif(mode = 0) {
        this.chain.add({ name: 'yadif', positional: [], named: { mode } });
        return this;
    }
    transpose(dir) {
        this.chain.add({ name: 'transpose', positional: [dir], named: {} });
        return this;
    }
    fade(type, startFrame, nbFrames) {
        this.chain.add({ name: 'fade', positional: [], named: { type, start_frame: startFrame, nb_frames: nbFrames } });
        return this;
    }
    thumbnail(n = 100) {
        this.chain.add({ name: 'thumbnail', positional: [n], named: {} });
        return this;
    }
    toString() { return this.chain.toString(); }
}
/**
 * Build a typed audio filter chain for use with .audioFilter().
 *
 * @example
 * const af = audioFilterChain()
 *   .loudnorm(-23, 7, -2)
 *   .highpass(80)
 *   .toString();
 */
export class AudioFilterChain {
    chain = new FilterChain();
    raw(str) { this.chain.raw(str); return this; }
    volume(v) {
        this.chain.add({ name: 'volume', positional: [String(v)], named: {} });
        return this;
    }
    loudnorm(i = -23, lra = 7, tp = -2) {
        this.chain.add({ name: 'loudnorm', positional: [], named: { i, lra, tp } });
        return this;
    }
    highpass(freq) {
        this.chain.add({ name: 'highpass', positional: [], named: { f: freq } });
        return this;
    }
    lowpass(freq) {
        this.chain.add({ name: 'lowpass', positional: [], named: { f: freq } });
        return this;
    }
    equalizer(freq, width, gain, widthType = 'o') {
        this.chain.add({ name: 'equalizer', positional: [], named: { f: freq, width, g: gain, width_type: widthType } });
        return this;
    }
    afade(type, startTime, duration) {
        this.chain.add({ name: 'afade', positional: [], named: { type, start_time: startTime, duration } });
        return this;
    }
    atempo(factor) {
        this.chain.add({ name: 'atempo', positional: [factor], named: {} });
        return this;
    }
    asetpts(expr) {
        this.chain.add({ name: 'asetpts', positional: [expr], named: {} });
        return this;
    }
    dynaudnorm(frameLen = 500, peak = 0.95) {
        this.chain.add({ name: 'dynaudnorm', positional: [], named: { f: frameLen, p: peak } });
        return this;
    }
    silencedetect(noise = '-60dB', duration = 2) {
        this.chain.add({ name: 'silencedetect', positional: [], named: { noise, duration } });
        return this;
    }
    aecho(inGain = 0.6, outGain = 0.3, delays = '1000', decays = '0.5') {
        this.chain.add({ name: 'aecho', positional: [], named: { in_gain: inGain, out_gain: outGain, delays, decays } });
        return this;
    }
    rubberband(tempo, pitch) {
        const named = {};
        if (tempo !== undefined)
            named['tempo'] = tempo;
        if (pitch !== undefined)
            named['pitch'] = pitch;
        this.chain.add({ name: 'rubberband', positional: [], named });
        return this;
    }
    toString() { return this.chain.toString(); }
}
/** Factory: create a video filter chain */
export function videoFilterChain() {
    return new VideoFilterChain();
}
/** Factory: create an audio filter chain */
export function audioFilterChain() {
    return new AudioFilterChain();
}
/** Factory: create a filter_complex graph builder */
export function filterGraph() {
    return new FilterGraph();
}
