// mediaforge/dist/esm/codecs/video.js
/**
 * Serialize X264Options to ffmpeg argument array.
 * Emits codec args that follow -c:v libx264.
 */
export function x264ToArgs(opts) {
    const args = ['-c:v', 'libx264'];
    if (opts.preset !== undefined)
        args.push('-preset', opts.preset);
    if (opts.tune !== undefined)
        args.push('-tune', opts.tune);
    if (opts.profile !== undefined)
        args.push('-profile:v', opts.profile);
    if (opts.crf !== undefined)
        args.push('-crf', String(opts.crf));
    if (opts.qp !== undefined)
        args.push('-qp', String(opts.qp));
    if (opts.bFrames !== undefined)
        args.push('-bf', String(opts.bFrames));
    if (opts.bPyramid !== undefined)
        args.push('-b_strategy', opts.bPyramid === 'none' ? '0' : '1');
    if (opts.aqMode !== undefined)
        args.push('-aq-mode', String(opts.aqMode));
    if (opts.aqStrength !== undefined)
        args.push('-aq-strength', String(opts.aqStrength));
    if (opts.weightp !== undefined)
        args.push('-weightp', String(opts.weightp));
    if (opts.refs !== undefined)
        args.push('-refs', String(opts.refs));
    if (opts.subq !== undefined)
        args.push('-subq', String(opts.subq));
    if (opts.me !== undefined)
        args.push('-me_method', opts.me);
    if (opts.meRange !== undefined)
        args.push('-me_range', String(opts.meRange));
    if (opts.trellis !== undefined)
        args.push('-trellis', String(opts.trellis));
    if (opts.deblock !== undefined)
        args.push('-deblock', opts.deblock);
    if (opts.nalHrd !== undefined)
        args.push('-nal-hrd', opts.nalHrd);
    if (opts.bitrate !== undefined)
        args.push('-b:v', `${opts.bitrate}k`);
    if (opts.maxrate !== undefined)
        args.push('-maxrate', `${opts.maxrate}k`);
    if (opts.bufsize !== undefined)
        args.push('-bufsize', `${opts.bufsize}k`);
    if (opts.qpMin !== undefined)
        args.push('-qmin', String(opts.qpMin));
    if (opts.qpMax !== undefined)
        args.push('-qmax', String(opts.qpMax));
    if (opts.keyintMax !== undefined)
        args.push('-g', String(opts.keyintMax));
    if (opts.keyintMin !== undefined)
        args.push('-keyint_min', String(opts.keyintMin));
    if (opts.scenecutThreshold !== undefined)
        args.push('-sc_threshold', String(opts.scenecutThreshold));
    if (opts.x264Params !== undefined)
        args.push('-x264-params', opts.x264Params);
    if (opts.pixFmt !== undefined)
        args.push('-pix_fmt', opts.pixFmt);
    return args;
}
/**
 * Serialize X265Options to ffmpeg argument array.
 */
export function x265ToArgs(opts) {
    const args = ['-c:v', 'libx265'];
    if (opts.preset !== undefined)
        args.push('-preset', opts.preset);
    if (opts.tune !== undefined)
        args.push('-tune', opts.tune);
    if (opts.profile !== undefined)
        args.push('-profile:v', opts.profile);
    if (opts.crf !== undefined)
        args.push('-crf', String(opts.crf));
    if (opts.qp !== undefined)
        args.push('-qp', String(opts.qp));
    if (opts.bitrate !== undefined)
        args.push('-b:v', `${opts.bitrate}k`);
    if (opts.maxrate !== undefined)
        args.push('-maxrate', `${opts.maxrate}k`);
    if (opts.bufsize !== undefined)
        args.push('-bufsize', `${opts.bufsize}k`);
    if (opts.keyintMax !== undefined)
        args.push('-g', String(opts.keyintMax));
    if (opts.bFrames !== undefined)
        args.push('-bf', String(opts.bFrames));
    if (opts.refs !== undefined)
        args.push('-refs', String(opts.refs));
    if (opts.dolbyVision === true)
        args.push('-dolbyvision', '1');
    if (opts.x265Params !== undefined)
        args.push('-x265-params', opts.x265Params);
    if (opts.pixFmt !== undefined)
        args.push('-pix_fmt', opts.pixFmt);
    return args;
}
/**
 * Serialize SvtAv1Options to ffmpeg argument array.
 */
export function svtav1ToArgs(opts) {
    const args = ['-c:v', 'libsvtav1'];
    if (opts.preset !== undefined)
        args.push('-preset', String(opts.preset));
    if (opts.crf !== undefined)
        args.push('-crf', String(opts.crf));
    if (opts.qp !== undefined)
        args.push('-qp', String(opts.qp));
    if (opts.bitrate !== undefined)
        args.push('-b:v', `${opts.bitrate}k`);
    if (opts.bFrames !== undefined)
        args.push('-bf', String(opts.bFrames));
    if (opts.keyintMax !== undefined)
        args.push('-g', String(opts.keyintMax));
    if (opts.enableSceneDetect !== undefined)
        args.push('-svtav1-params', `scd=${opts.enableSceneDetect ? 1 : 0}`);
    if (opts.dolbyVision === true)
        args.push('-dolbyvision', '1');
    if (opts.svtav1Params !== undefined)
        args.push('-svtav1-params', opts.svtav1Params);
    if (opts.pixFmt !== undefined)
        args.push('-pix_fmt', opts.pixFmt);
    return args;
}
/**
 * Serialize Vp9Options to ffmpeg argument array.
 */
export function vp9ToArgs(opts) {
    const args = ['-c:v', 'libvpx-vp9'];
    if (opts.bitrate !== undefined)
        args.push('-b:v', `${opts.bitrate}k`);
    if (opts.minrate !== undefined)
        args.push('-minrate', `${opts.minrate}k`);
    if (opts.maxrate !== undefined)
        args.push('-maxrate', `${opts.maxrate}k`);
    if (opts.crf !== undefined)
        args.push('-crf', String(opts.crf));
    if (opts.quality !== undefined)
        args.push('-quality', opts.quality);
    if (opts.cpuUsed !== undefined)
        args.push('-cpu-used', String(opts.cpuUsed));
    if (opts.tileColumns !== undefined)
        args.push('-tile-columns', String(opts.tileColumns));
    if (opts.tileRows !== undefined)
        args.push('-tile-rows', String(opts.tileRows));
    if (opts.rowMt !== undefined)
        args.push('-row-mt', opts.rowMt ? '1' : '0');
    if (opts.aqMode !== undefined)
        args.push('-aq-mode', String(opts.aqMode));
    if (opts.lagInFrames !== undefined)
        args.push('-lag-in-frames', String(opts.lagInFrames));
    if (opts.autoAltRef !== undefined)
        args.push('-auto-alt-ref', opts.autoAltRef ? '1' : '0');
    if (opts.arnrMaxFrames !== undefined)
        args.push('-arnr-maxframes', String(opts.arnrMaxFrames));
    if (opts.pass !== undefined)
        args.push('-pass', String(opts.pass));
    if (opts.passlogfile !== undefined)
        args.push('-passlogfile', opts.passlogfile);
    if (opts.keyintMax !== undefined)
        args.push('-g', String(opts.keyintMax));
    if (opts.pixFmt !== undefined)
        args.push('-pix_fmt', opts.pixFmt);
    return args;
}
// Alias for camelCase import compatibility
export { svtav1ToArgs as svtAv1ToArgs };
export function proResToArgs(opts = {}, encoder = 'prores_ks') {
    const args = ['-c:v', encoder];
    if (opts.profile !== undefined)
        args.push('-profile:v', String(opts.profile));
    if (opts.bits !== undefined)
        args.push('-bits_per_raw_sample', String(opts.bits));
    if (opts.vendor !== undefined)
        args.push('-vendor', opts.vendor);
    if (opts.alphaQuality !== undefined)
        args.push('-alpha_quality', String(opts.alphaQuality));
    if (opts.mbs !== undefined)
        args.push('-mbs_per_slice', opts.mbs === 'adaptive' ? '0' : '8');
    return args;
}
export function dnxhdToArgs(opts = {}) {
    const args = ['-c:v', 'dnxhd'];
    if (opts.profile !== undefined)
        args.push('-profile:v', opts.profile);
    if (opts.bitrate !== undefined)
        args.push('-b:v', `${opts.bitrate}k`);
    if (opts.pixFmt !== undefined)
        args.push('-pix_fmt', opts.pixFmt);
    return args;
}
export function mjpegToArgs(opts = {}) {
    const args = ['-c:v', 'mjpeg'];
    if (opts.qscale !== undefined)
        args.push('-q:v', String(opts.qscale));
    if (opts.huffman !== undefined)
        args.push('-huffman', opts.huffman);
    if (opts.pixFmt !== undefined)
        args.push('-pix_fmt', opts.pixFmt);
    return args;
}
export function mpeg2ToArgs(opts = {}) {
    const args = ['-c:v', 'mpeg2video'];
    if (opts.bitrate !== undefined)
        args.push('-b:v', `${opts.bitrate}k`);
    if (opts.maxrate !== undefined)
        args.push('-maxrate', `${opts.maxrate}k`);
    if (opts.bufsize !== undefined)
        args.push('-bufsize', `${opts.bufsize}k`);
    if (opts.gopSize !== undefined)
        args.push('-g', String(opts.gopSize));
    if (opts.profile !== undefined)
        args.push('-profile:v', opts.profile);
    if (opts.level !== undefined)
        args.push('-level', opts.level);
    if (opts.interlaced === true)
        args.push('-flags', '+ildct+ilme');
    return args;
}
export function mpeg4ToArgs(opts = {}, encoder = 'mpeg4') {
    const args = ['-c:v', encoder];
    if (opts.bitrate !== undefined)
        args.push('-b:v', `${opts.bitrate}k`);
    if (opts.maxrate !== undefined)
        args.push('-maxrate', `${opts.maxrate}k`);
    if (opts.qscale !== undefined)
        args.push('-q:v', String(opts.qscale));
    if (opts.gopSize !== undefined)
        args.push('-g', String(opts.gopSize));
    if (opts.bFrames !== undefined)
        args.push('-bf', String(opts.bFrames));
    if (opts.me !== undefined)
        args.push('-me_method', opts.me);
    return args;
}
export function vp8ToArgs(opts = {}) {
    const args = ['-c:v', 'libvpx'];
    if (opts.bitrate !== undefined)
        args.push('-b:v', `${opts.bitrate}k`);
    if (opts.minrate !== undefined)
        args.push('-minrate', `${opts.minrate}k`);
    if (opts.maxrate !== undefined)
        args.push('-maxrate', `${opts.maxrate}k`);
    if (opts.crf !== undefined)
        args.push('-crf', String(opts.crf));
    if (opts.cpuUsed !== undefined)
        args.push('-cpu-used', String(opts.cpuUsed));
    if (opts.quality !== undefined)
        args.push('-quality', opts.quality);
    if (opts.keyintMax !== undefined)
        args.push('-g', String(opts.keyintMax));
    return args;
}
export function theoraToArgs(opts = {}) {
    const args = ['-c:v', 'libtheora'];
    if (opts.qscale !== undefined)
        args.push('-q:v', String(opts.qscale));
    else if (opts.bitrate !== undefined)
        args.push('-b:v', `${opts.bitrate}k`);
    return args;
}
export function ffv1ToArgs(opts = {}) {
    const args = ['-c:v', 'ffv1'];
    if (opts.version !== undefined)
        args.push('-level', String(opts.version));
    if (opts.coder !== undefined)
        args.push('-coder', String(opts.coder));
    if (opts.context !== undefined)
        args.push('-context', String(opts.context));
    if (opts.slices !== undefined)
        args.push('-slices', String(opts.slices));
    if (opts.sliceCrc !== undefined)
        args.push('-slicecrc', opts.sliceCrc ? '1' : '0');
    return args;
}
