// mediaforge/dist/cjs/helpers/streams.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.pipeThrough = pipeThrough;
exports.streamOutput = streamOutput;
exports.streamToFile = streamToFile;
exports.buildPipeThroughArgs = buildPipeThroughArgs;
exports.buildStreamOutputArgs = buildStreamOutputArgs;
const child_process_1 = require("child_process");
const fs_1 = require("fs");
const stream_1 = require("stream");
const events_js_1 = require("../process/events.js");
const progress_js_1 = require("../process/progress.js");
const spawn_js_1 = require("../process/spawn.js");
const binary_js_1 = require("../utils/binary.js");
const readline_1 = require("readline");
/**
 * Pipe data through ffmpeg — input from a stream, output to a stream.
 *
 * @example
 * // Transcode a readable stream to a writable stream
 * const proc = pipeThrough({
 *   inputFormat: 'mp4',
 *   outputArgs: ['-c:v', 'libx264', '-c:a', 'aac'],
 *   outputFormat: 'mp4',
 * });
 * fsReadStream.pipe(proc.stdin!);
 * proc.stdout.pipe(fsWriteStream);
 * await new Promise((res, rej) => {
 *   proc.emitter.on('end', res);
 *   proc.emitter.on('error', rej);
 * });
 *
 * @example
 * // Pass a readable stream directly
 * const proc = pipeThrough({ inputStream: myReadable, outputFormat: 'ogg' });
 * proc.stdout.pipe(response); // stream to HTTP response
 */
function pipeThrough(opts) {
    const { inputStream, inputFormat, outputFormat, parseProgress = false, binary = (0, binary_js_1.resolveBinary)(), } = opts;
    let { outputArgs = [] } = opts;
    // MP4/MOV written to a pipe cannot seek back to write the moov atom.
    // Automatically inject fragmented-MP4 flags so the output is streamable.
    // Only injected when the user has not already supplied -movflags themselves.
    const pipedContainerNeedsFragmentation = outputFormat === 'mp4' || outputFormat === 'mov';
    const userAlreadySetMovflags = outputArgs.includes('-movflags');
    if (pipedContainerNeedsFragmentation && !userAlreadySetMovflags) {
        outputArgs = [...outputArgs, '-movflags', 'frag_keyframe+empty_moov+default_base_moof'];
    }
    const args = ['-y'];
    // MP4/MOV piped as INPUT: the moov atom is normally at the end of the file,
    // so FFmpeg cannot find codec parameters when reading from a non-seekable pipe.
    // Increase analyzeduration and probesize so FFmpeg buffers enough data to
    // detect stream parameters before giving up with "unspecified pixel format".
    const pipedInputNeedsProbe = inputFormat === 'mp4' || inputFormat === 'mov' || inputFormat === 'm4v';
    if (pipedInputNeedsProbe) {
        // Large probe window + genpts for recovery when moov atom is at end of file
        args.push('-analyzeduration', '100M', '-probesize', '100M');
        args.push('-fflags', '+genpts');
    }
    if (inputFormat)
        args.push('-f', inputFormat);
    args.push('-i', 'pipe:0');
    args.push(...outputArgs);
    if (outputFormat)
        args.push('-f', outputFormat);
    args.push('pipe:1');
    const child = (0, child_process_1.spawn)(binary, args, { stdio: ['pipe', 'pipe', 'pipe'] });
    const emitter = new events_js_1.FFmpegEmitter();
    const stderrLines = [];
    const progressParser = parseProgress
        ? new progress_js_1.ProgressParser(info => emitter.emit('progress', info))
        : null;
    emitter.emit('start', args);
    if (child.stderr) {
        const rl = (0, readline_1.createInterface)({ input: child.stderr, crlfDelay: Infinity });
        rl.on('line', line => {
            stderrLines.push(line);
            emitter.emit('stderr', line);
            progressParser?.push(line);
        });
    }
    child.on('close', (code, signal) => {
        if (code === 0 || (code === null && signal === null)) {
            emitter.emit('end');
        }
        else {
            emitter.emit('error', new spawn_js_1.FFmpegSpawnError(code, signal, stderrLines.join('\n')));
        }
    });
    child.on('error', err => emitter.emit('error', err));
    // Pipe inputStream into stdin automatically if provided
    if (inputStream && child.stdin) {
        inputStream.pipe(child.stdin);
        inputStream.on('error', err => child.stdin?.destroy(err));
    }
    return {
        emitter,
        stdin: child.stdin,
        stdout: child.stdout,
        kill(signal = 'SIGTERM') { child.kill(signal); },
    };
}
function streamOutput(opts) {
    const { input, outputArgs = [], outputFormat, seekInput, binary = (0, binary_js_1.resolveBinary)(), } = opts;
    const args = ['-y'];
    if (seekInput !== undefined)
        args.push('-ss', String(seekInput));
    args.push('-i', input);
    args.push(...outputArgs);
    args.push('-f', outputFormat, 'pipe:1');
    const child = (0, child_process_1.spawn)(binary, args, { stdio: ['ignore', 'pipe', 'pipe'] });
    const pass = new stream_1.PassThrough();
    child.stdout?.pipe(pass);
    child.on('error', err => pass.destroy(err));
    child.on('close', (code) => {
        if (code !== 0 && code !== null) {
            pass.destroy(new Error(`ffmpeg exited with code ${code}`));
        }
    });
    return pass;
}
function streamToFile(opts) {
    const { stream, inputFormat, output, outputArgs = [], binary = (0, binary_js_1.resolveBinary)(), } = opts;
    // Non-seekable pipes fail for MP4/MOV because the moov atom is at the file end.
    // Buffer the entire stream to a temp file first, then run FFmpeg on that file.
    const needsTempFile = inputFormat === 'mp4' || inputFormat === 'mov' || inputFormat === 'm4v';
    if (needsTempFile) {
        const tmpPath = `${output}.tmp_${Date.now()}.${inputFormat ?? 'tmp'}`;
        return new Promise((resolve, reject) => {
            const ws = (0, fs_1.createWriteStream)(tmpPath);
            stream.pipe(ws);
            stream.on('error', reject);
            ws.on('error', reject);
            ws.on('close', () => {
                const args = ['-y'];
                if (inputFormat)
                    args.push('-f', inputFormat);
                args.push('-i', tmpPath, ...outputArgs, output);
                const child = (0, child_process_1.spawn)(binary, args, { stdio: ['ignore', 'ignore', 'pipe'] });
                const stderrLines = [];
                if (child.stderr) {
                    const rl = (0, readline_1.createInterface)({ input: child.stderr, crlfDelay: Infinity });
                    rl.on('line', (line) => stderrLines.push(line));
                }
                child.on('close', (code, signal) => {
                    try {
                        (0, fs_1.unlinkSync)(tmpPath);
                    }
                    catch { /**/ }
                    if (code === 0 || (code === null && signal === null))
                        resolve();
                    else
                        reject(new spawn_js_1.FFmpegSpawnError(code, signal, stderrLines.join('\n')));
                });
                child.on('error', (err) => { try {
                    (0, fs_1.unlinkSync)(tmpPath);
                }
                catch { /**/ } reject(err); });
            });
        });
    }
    // All other formats: pipe directly to stdin
    return new Promise((resolve, reject) => {
        const args = ['-y'];
        if (inputFormat)
            args.push('-f', inputFormat);
        args.push('-i', 'pipe:0', ...outputArgs, output);
        const child = (0, child_process_1.spawn)(binary, args, { stdio: ['pipe', 'ignore', 'pipe'] });
        const _emitter = new events_js_1.FFmpegEmitter();
        const stderrLines = [];
        if (child.stderr) {
            const rl = (0, readline_1.createInterface)({ input: child.stderr, crlfDelay: Infinity });
            rl.on('line', (line) => stderrLines.push(line));
        }
        stream.pipe(child.stdin);
        stream.on('error', (err) => child.stdin?.destroy(err));
        child.on('close', (code, signal) => {
            if (code === 0 || (code === null && signal === null))
                resolve();
            else
                reject(new spawn_js_1.FFmpegSpawnError(code, signal, stderrLines.join('\n')));
        });
        child.on('error', reject);
    });
}
// ─── Arg builders ─────────────────────────────────────────────────────────────
function buildPipeThroughArgs(inputFormat, outputArgs, outputFormat) {
    const args = ['-y'];
    // MP4/MOV piped as INPUT: moov atom is at end of file — buffer more data
    const pipedInputNeedsProbe = inputFormat === 'mp4' || inputFormat === 'mov' || inputFormat === 'm4v';
    if (pipedInputNeedsProbe) {
        // Large probe window + genpts for recovery when moov atom is at end of file
        args.push('-analyzeduration', '100M', '-probesize', '100M');
        args.push('-fflags', '+genpts');
    }
    if (inputFormat)
        args.push('-f', inputFormat);
    args.push('-i', 'pipe:0');
    // Auto-inject fragmented-MP4 flags when piping to MP4/MOV (no seek available)
    const pipedContainerNeedsFragmentation = outputFormat === 'mp4' || outputFormat === 'mov';
    const userAlreadySetMovflags = outputArgs.includes('-movflags');
    const effectiveOutputArgs = pipedContainerNeedsFragmentation && !userAlreadySetMovflags
        ? [...outputArgs, '-movflags', 'frag_keyframe+empty_moov+default_base_moof']
        : outputArgs;
    args.push(...effectiveOutputArgs);
    if (outputFormat)
        args.push('-f', outputFormat);
    args.push('pipe:1');
    return args;
}
function buildStreamOutputArgs(input, outputArgs, outputFormat, seekInput) {
    const args = ['-y'];
    if (seekInput !== undefined)
        args.push('-ss', String(seekInput));
    args.push('-i', input, ...outputArgs, '-f', outputFormat, 'pipe:1');
    return args;
}
