// mediaforge/dist/cjs/probe/ffprobe.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ProbeError = void 0;
exports.probe = probe;
exports.probeAsync = probeAsync;
exports.parseFrameRate = parseFrameRate;
exports.parseDuration = parseDuration;
exports.parseBitrate = parseBitrate;
exports.getVideoStreams = getVideoStreams;
exports.getAudioStreams = getAudioStreams;
exports.getSubtitleStreams = getSubtitleStreams;
exports.getDefaultVideoStream = getDefaultVideoStream;
exports.getDefaultAudioStream = getDefaultAudioStream;
exports.getMediaDuration = getMediaDuration;
exports.durationToMicroseconds = durationToMicroseconds;
exports.summarizeVideoStream = summarizeVideoStream;
exports.summarizeAudioStream = summarizeAudioStream;
exports.getStreamLanguage = getStreamLanguage;
exports.findStreamByLanguage = findStreamByLanguage;
exports.formatDuration = formatDuration;
exports.isHdr = isHdr;
exports.isInterlaced = isInterlaced;
exports.getChapterList = getChapterList;
const child_process_1 = require("child_process");
const binary_js_1 = require("../utils/binary.js");
/**
 * Run ffprobe on a file and return fully typed JSON output.
 * Synchronous version — use `probeAsync` for non-blocking operation.
 *
 * @example
 * const info = probe('input.mp4');
 * console.log(info.format?.duration); // "120.042000"
 * console.log(info.streams[0].codec_name); // "h264"
 */
function probe(filePath, opts = {}) {
    const binary = (0, binary_js_1.resolveProbe)(opts.binary);
    const args = buildProbeArgs(filePath, opts);
    let output;
    try {
        output = (0, child_process_1.execFileSync)(binary, args, {
            stdio: ['ignore', 'pipe', 'pipe'],
            encoding: 'utf8',
            timeout: opts.timeout ?? 30000,
        });
    }
    catch (err) {
        const msg = err.message ?? String(err);
        throw new ProbeError(filePath, msg);
    }
    return parseProbeOutput(output, filePath);
}
/**
 * Async version of probe — non-blocking, suitable for server use.
 */
function probeAsync(filePath, opts = {}) {
    return new Promise((resolve, reject) => {
        const binary = (0, binary_js_1.resolveProbe)(opts.binary);
        const args = buildProbeArgs(filePath, opts);
        const child = (0, child_process_1.spawn)(binary, args, { stdio: ['ignore', 'pipe', 'pipe'] });
        let stdout = '';
        let stderr = '';
        child.stdout?.on('data', (chunk) => { stdout += chunk.toString(); });
        child.stderr?.on('data', (chunk) => { stderr += chunk.toString(); });
        child.on('close', (code) => {
            if (code !== 0) {
                reject(new ProbeError(filePath, stderr.slice(-1000)));
                return;
            }
            try {
                resolve(parseProbeOutput(stdout, filePath));
            }
            catch (e) {
                reject(e);
            }
        });
        child.on('error', (err) => reject(new ProbeError(filePath, err.message)));
    });
}
// ─── Internal helpers ─────────────────────────────────────────────────────────
function buildProbeArgs(filePath, opts) {
    const showArgs = ['-show_format', '-show_streams'];
    if (opts.chapters !== false)
        showArgs.push('-show_chapters');
    return [
        '-v', 'quiet',
        '-print_format', 'json',
        ...showArgs,
        ...(opts.extraArgs ?? []),
        filePath,
    ];
}
function parseProbeOutput(output, filePath) {
    let parsed;
    try {
        parsed = JSON.parse(output);
    }
    catch {
        throw new ProbeError(filePath, `Failed to parse ffprobe JSON output: ${output.slice(0, 200)}`);
    }
    if (parsed === null || typeof parsed !== 'object') {
        throw new ProbeError(filePath, 'ffprobe returned non-object JSON');
    }
    const obj = parsed;
    const result = {
        streams: obj['streams'] ?? [],
    };
    const fmt = obj['format'];
    if (fmt !== undefined)
        result.format = fmt;
    const chs = obj['chapters'];
    if (chs !== undefined)
        result.chapters = chs;
    return result;
}
// ─── Error class ──────────────────────────────────────────────────────────────
class ProbeError extends Error {
    filePath;
    detail;
    constructor(filePath, detail) {
        super(`ffprobe failed for "${filePath}": ${detail}`);
        this.filePath = filePath;
        this.detail = detail;
        this.name = 'ProbeError';
    }
}
exports.ProbeError = ProbeError;
// ─── Derived helpers ──────────────────────────────────────────────────────────
/**
 * Parse an ffprobe fractional frame rate string.
 * e.g. "30000/1001" → { num:30000, den:1001, value:29.97 }
 * e.g. "25/1"       → { num:25,    den:1,    value:25 }
 *
 * @returns ParsedFrameRate object, or null if the string is absent/invalid.
 */
function parseFrameRate(frStr) {
    if (frStr === undefined || frStr === '' || frStr === '0/0')
        return null;
    const parts = frStr.split('/');
    if (parts.length !== 2)
        return null;
    const num = parseInt(parts[0] ?? '0', 10);
    const den = parseInt(parts[1] ?? '1', 10);
    if (den === 0)
        return null;
    return { num, den, value: num / den };
}
/**
 * Parse a duration string from ffprobe (e.g. "120.042000") to a number.
 * Returns null if the value is absent or "N/A".
 */
function parseDuration(durStr) {
    if (durStr === undefined || durStr === 'N/A' || durStr === '')
        return null;
    const n = parseFloat(durStr);
    return isNaN(n) ? null : n;
}
/**
 * Parse a bitrate string (e.g. "4200000") to a number.
 */
function parseBitrate(brStr) {
    if (brStr === undefined || brStr === 'N/A' || brStr === '')
        return null;
    const n = parseInt(brStr, 10);
    return isNaN(n) ? null : n;
}
/**
 * Get all video streams from a ProbeResult.
 */
function getVideoStreams(result) {
    return result.streams.filter((s) => s.codec_type === 'video');
}
/**
 * Get all audio streams from a ProbeResult.
 */
function getAudioStreams(result) {
    return result.streams.filter((s) => s.codec_type === 'audio');
}
/**
 * Get all subtitle streams from a ProbeResult.
 */
function getSubtitleStreams(result) {
    return result.streams.filter((s) => s.codec_type === 'subtitle');
}
/**
 * Get the first video stream, or null if none exists.
 */
function getDefaultVideoStream(result) {
    return getVideoStreams(result)[0] ?? null;
}
/**
 * Get the first audio stream, or null if none exists.
 */
function getDefaultAudioStream(result) {
    return getAudioStreams(result)[0] ?? null;
}
/**
 * Return the total duration of the media in seconds.
 * Prefers format.duration, falls back to the first stream with a duration.
 */
function getMediaDuration(result) {
    const fromFormat = parseDuration(result.format?.duration);
    if (fromFormat !== null)
        return fromFormat;
    for (const stream of result.streams) {
        const d = parseDuration(stream.duration);
        if (d !== null)
            return d;
    }
    return null;
}
/**
 * Convert total duration in seconds to microseconds (for progress calculation).
 */
function durationToMicroseconds(seconds) {
    return Math.round(seconds * 1_000_000);
}
/**
 * Build a human-readable summary of a video stream.
 */
function summarizeVideoStream(s) {
    const fps = parseFrameRate(s.avg_frame_rate ?? s.r_frame_rate);
    const vss = {
        index: s.index,
        codec: s.codec_name ?? 'unknown',
        width: s.width ?? 0,
        height: s.height ?? 0,
        fps: fps?.value ?? 0,
        pixFmt: s.pix_fmt ?? 'unknown',
        durationSec: parseDuration(s.duration),
        bitrateBps: parseBitrate(s.bit_rate),
    };
    if (s.profile !== undefined)
        vss.profile = s.profile;
    if (s.color_space !== undefined)
        vss.colorSpace = s.color_space;
    return vss;
}
/**
 * Build a human-readable summary of an audio stream.
 */
function summarizeAudioStream(s) {
    return {
        index: s.index,
        codec: s.codec_name ?? 'unknown',
        sampleRate: parseInt(s.sample_rate ?? '0', 10),
        channels: s.channels ?? 0,
        channelLayout: s.channel_layout ?? 'unknown',
        durationSec: parseDuration(s.duration),
        bitrateBps: parseBitrate(s.bit_rate),
    };
}
/**
 * Get a stream's language tag, or null if not set.
 */
function getStreamLanguage(s) {
    return s.tags?.['language'] ?? null;
}
/**
 * Find a stream by language tag (case-insensitive).
 * Returns the first match, or null.
 */
function findStreamByLanguage(result, lang, codecType) {
    const lowerLang = lang.toLowerCase();
    for (const s of result.streams) {
        if (codecType !== undefined && s.codec_type !== codecType)
            continue;
        const sl = getStreamLanguage(s)?.toLowerCase();
        if (sl === lowerLang)
            return s;
    }
    return null;
}
/**
 * Format a duration in seconds to a human-readable HH:MM:SS.mmm string.
 */
function formatDuration(seconds) {
    const h = Math.floor(seconds / 3600);
    const m = Math.floor((seconds % 3600) / 60);
    const s = seconds % 60;
    return `${String(h).padStart(2, '0')}:${String(m).padStart(2, '0')}:${s.toFixed(3).padStart(6, '0')}`;
}
/**
 * Check if a file appears to be HDR (BT.2020 primaries or PQ/HLG transfer).
 */
function isHdr(result) {
    const vs = getDefaultVideoStream(result);
    if (vs === null)
        return false;
    const hdrPrimaries = ['bt2020'];
    const hdrTransfer = ['smpte2084', 'arib-std-b67', 'smpte428', 'bt2020-10', 'bt2020-12'];
    return (hdrPrimaries.includes(vs.color_primaries ?? '') ||
        hdrTransfer.includes(vs.color_transfer ?? ''));
}
/**
 * Check if a file's video stream is interlaced.
 */
function isInterlaced(result) {
    const vs = getDefaultVideoStream(result);
    if (vs === null)
        return false;
    const fo = vs.field_order;
    return fo !== undefined && fo !== 'progressive' && fo !== 'unknown';
}
/**
 * Get all chapter titles and times as a simple array.
 */
function getChapterList(result) {
    return (result.chapters ?? []).map((ch) => ({
        title: ch.tags?.['title'] ?? `Chapter ${ch.id ?? '?'}`,
        startSec: parseDuration(ch.start_time) ?? 0,
        endSec: parseDuration(ch.end_time) ?? 0,
    }));
}
